{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "98f3bc33",
   "metadata": {},
   "source": [
    "Fonctionnement du script\n",
    "------------\n",
    "Le script permet de créer dans un fichier FASTA toutes les mutations qui s'appliquent pour chaque transcrit pour chaque gène contenu dans la base de données de COSMICS, dans le but d'obtenir une prédiction du CDS avant et après la mutation à l'aide de FOMONET.\n",
    "\n",
    "Chaque cellule doit, à tour de rôle et dans l'ordre, être roulée.\n",
    "\n",
    "À quelques endroits dans le script, on retrouve des prints qui indique si tout fonctionne bien ou s'il y a des erreurs.\n",
    "\n",
    "À la fin, c'est le fichier '../result/all_transcripts_with_mutation.fasta' qui doit être rouler dans FOMOnet. Il contient tous les transcrits de chaque gène à l'état WILD-TYPE en plus de tous les transcrits des gènes contenu dans COSMICS avec chacune des mutations. (Il y a toujours une seule mutation effectuée dans chaque transcrit (PAS DE COMBINAISON DE MUTATION))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c62b593d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'PRDM16': 'ENSG00000142611', 'THRAP3': 'ENSG00000054118', 'CDKN2C': 'ENSG00000123080', 'TENT5C': 'ENSG00000183508', 'TPR': 'ENSG00000047410', 'PRRX1': 'ENSG00000116132', 'ARHGEF10L': 'ENSG00000074964', 'ID3': 'ENSG00000117318', 'ATP1A1': 'ENSG00000163399', 'H3F3A': 'ENSG00000163041', 'ARNT': 'ENSG00000143437', 'MPL': 'ENSG00000117400', 'ASPM': 'ENSG00000066279', 'AKT3': 'ENSG00000117020', 'PRCC': 'ENSG00000143294', 'TNFRSF14': 'ENSG00000157873', 'RBM15': 'ENSG00000162775', 'STIL': 'ENSG00000123473', 'NRAS': 'ENSG00000213281', 'MLLT11': 'ENSG00000213190', 'FH': 'ENSG00000091483', 'BCL9': 'ENSG00000116128', 'ELF3': 'ENSG00000163435', 'RGS7': 'ENSG00000182901', 'MDM4': 'ENSG00000198625', 'NOTCH2': 'ENSG00000134250', 'PIK3R3': 'ENSG00000117461', 'SETDB1': 'ENSG00000143379', 'PRDM2': 'ENSG00000116731', 'TAL1': 'ENSG00000162367', 'FCGR2B': 'ENSG00000072694', 'SFPQ': 'ENSG00000116560', 'FCRL4': 'ENSG00000163518', 'DDR2': 'ENSG00000162733', 'LCK': 'ENSG00000182866', 'S100A7': 'ENSG00000143556', 'JAK1': 'ENSG00000162434', 'PBX1': 'ENSG00000185630', 'FUBP1': 'ENSG00000162613', 'SPEN': 'ENSG00000065526', 'SKI': 'ENSG00000157933', 'ABL2': 'ENSG00000143322', 'TPM3': 'ENSG00000143549', 'EPS15': 'ENSG00000085832', 'BTG2': 'ENSG00000159388', 'RPL22': 'ENSG00000116251', 'PDE4DIP': 'ENSG00000178104', 'NTRK1': 'ENSG00000198400', 'JUN': 'ENSG00000177606', 'CAMTA1': 'ENSG00000171735', 'TRIM33': 'ENSG00000197323', 'CSF3R': 'ENSG00000119535', 'MYCL': 'ENSG00000116990', 'PAX7': 'ENSG00000009709', 'CDC73': 'ENSG00000134371', 'ARID1A': 'ENSG00000117713', 'LMNA': 'ENSG00000160789', 'MUTYH': 'ENSG00000132781', 'SDHC': 'ENSG00000143252', 'MTOR': 'ENSG00000198793', 'ELK4': 'ENSG00000158711', 'SLC45A3': 'ENSG00000158715', 'SDHB': 'ENSG00000117118', 'CASP9': 'ENSG00000132906', 'PTPRC': 'ENSG00000081237', 'EPAS1': 'ENSG00000116016', 'BCL11A': 'ENSG00000119866', 'ALK': 'ENSG00000171094', 'EML4': 'ENSG00000143924', 'LRP1B': 'ENSG00000168702', 'BARD1': 'ENSG00000138376', 'ACKR3': 'ENSG00000144476', 'ATIC': 'ENSG00000138363', 'RGPD3': 'ENSG00000153165', 'PAX3': 'ENSG00000135903', 'FEV': 'ENSG00000163497', 'ACVR2A': 'ENSG00000121989', 'IDH1': 'ENSG00000138413', 'SF3B1': 'ENSG00000115524', 'CTNNA2': 'ENSG00000066032', 'ACSL3': 'ENSG00000123983', 'DCTN1': 'ENSG00000204843', 'PCBP1': 'ENSG00000169564', 'ACVR1': 'ENSG00000115170', 'HOXD13': 'ENSG00000128714', 'HOXD11': 'ENSG00000128713', 'AFF3': 'ENSG00000144218', 'FBXO11': 'ENSG00000138081', 'NCOA1': 'ENSG00000084676', 'CUL3': 'ENSG00000036257', 'DNMT3A': 'ENSG00000119772', 'CASP8': 'ENSG00000064012', 'CXCR4': 'ENSG00000121966', 'ITGAV': 'ENSG00000138448', 'RANBP2': 'ENSG00000153201', 'WDCP': 'ENSG00000163026', 'REL': 'ENSG00000162924', 'NFE2L2': 'ENSG00000116044', 'BIRC6': 'ENSG00000115760', 'MYCN': 'ENSG00000134323', 'ERCC3': 'ENSG00000163161', 'PAX8': 'ENSG00000125618', 'ERBB4': 'ENSG00000178568', 'SIX2': 'ENSG00000170577', 'STRN': 'ENSG00000115808', 'MSH6': 'ENSG00000116062', 'ASXL2': 'ENSG00000143970', 'COL3A1': 'ENSG00000168542', 'XPO1': 'ENSG00000082898', 'MSH2': 'ENSG00000095002', 'TMEM127': 'ENSG00000135956', 'CREB1': 'ENSG00000118260', 'CD28': 'ENSG00000178562', 'PMS1': 'ENSG00000064933', 'EPHA3': 'ENSG00000044524', 'STAG1': 'ENSG00000118007', 'GMPS': 'ENSG00000163655', 'MECOM': 'ENSG00000085276', 'MAP3K13': 'ENSG00000073803', 'ATR': 'ENSG00000175054', 'MB21D2': 'ENSG00000180611', 'ETV5': 'ENSG00000244405', 'RPN1': 'ENSG00000163902', 'CNBP': 'ENSG00000169714', 'IGF2BP2': 'ENSG00000073792', 'TFG': 'ENSG00000114354', 'WWTR1': 'ENSG00000018408', 'EIF4A2': 'ENSG00000156976', 'BAP1': 'ENSG00000163930', 'GSK3B': 'ENSG00000082701', 'TP63': 'ENSG00000073282', 'RHOA': 'ENSG00000067560', 'NCKIPSD': 'ENSG00000213672', 'PIK3CA': 'ENSG00000121879', 'PPARG': 'ENSG00000132170', 'FANCD2': 'ENSG00000144554', 'POLQ': 'ENSG00000051341', 'FOXP1': 'ENSG00000114861', 'SRGAP3': 'ENSG00000196220', 'SETD2': 'ENSG00000181555', 'FBLN2': 'ENSG00000163520', 'TFRC': 'ENSG00000072274', 'RAF1': 'ENSG00000132155', 'CACNA1D': 'ENSG00000157388', 'TBL1XR1': 'ENSG00000177565', 'MUC4': 'ENSG00000145113', 'ROBO2': 'ENSG00000185008', 'PBRM1': 'ENSG00000163939', 'FHIT': 'ENSG00000189283', 'GATA2': 'ENSG00000179348', 'MITF': 'ENSG00000187098', 'MYD88': 'ENSG00000172936', 'BCL6': 'ENSG00000113916', 'MLH1': 'ENSG00000076242', 'TGFBR2': 'ENSG00000163513', 'CTNNB1': 'ENSG00000168036', 'SOX2': 'ENSG00000181449', 'VHL': 'ENSG00000134086', 'PIK3CB': 'ENSG00000051382', 'CCR4': 'ENSG00000183813', 'XPC': 'ENSG00000154767', 'SLC34A2': 'ENSG00000157765', 'PTPN13': 'ENSG00000163629', 'IL2': 'ENSG00000109471', 'RAP1GDS1': 'ENSG00000138698', 'FAT4': 'ENSG00000196159', 'TET2': 'ENSG00000168769', 'NSD2': 'ENSG00000109685', 'LEF1': 'ENSG00000138795', 'AFF1': 'ENSG00000172493', 'KIT': 'ENSG00000157404', 'FAT1': 'ENSG00000083857', 'PHOX2B': 'ENSG00000109132', 'TEC': 'ENSG00000135605', 'RHOH': 'ENSG00000168421', 'CASP3': 'ENSG00000164305', 'FIP1L1': 'ENSG00000145216', 'FBXW7': 'ENSG00000109670', 'FGFR3': 'ENSG00000068078', 'N4BP2': 'ENSG00000078177', 'PDGFRA': 'ENSG00000134853', 'CHIC2': 'ENSG00000109220', 'CDH10': 'ENSG00000040731', 'GOLPH3': 'ENSG00000113384', 'IL7R': 'ENSG00000168685', 'RAD50': 'ENSG00000113522', 'CTNNA1': 'ENSG00000044115', 'AFF4': 'ENSG00000072364', 'ARHGAP26': 'ENSG00000145819', 'LIFR': 'ENSG00000113594', 'CSF1R': 'ENSG00000182578', 'PDGFRB': 'ENSG00000113721', 'TLX3': 'ENSG00000164438', 'NPM1': 'ENSG00000181163', 'DROSHA': 'ENSG00000113360', 'RAD17': 'ENSG00000152942', 'FGFR4': 'ENSG00000160867', 'SDHA': 'ENSG00000073578', 'CD74': 'ENSG00000019582', 'MAP3K1': 'ENSG00000095015', 'ITK': 'ENSG00000113263', 'PIK3R1': 'ENSG00000145675', 'PWWP2A': 'ENSG00000170234', 'IL6ST': 'ENSG00000134352', 'TERT': 'ENSG00000164362', 'FLT4': 'ENSG00000037280', 'CTNND2': 'ENSG00000169862', 'EBF1': 'ENSG00000164330', 'NSD1': 'ENSG00000165671', 'APC': 'ENSG00000134982', 'SUB1': 'ENSG00000113387', 'TRIM27': 'ENSG00000204713', 'PRDM1': 'ENSG00000057657', 'BCLAF1': 'ENSG00000029363', 'BMP5': 'ENSG00000112175', 'GOPC': 'ENSG00000047932', 'FOXO3': 'ENSG00000118689', 'RSPO3': 'ENSG00000146374', 'LATS1': 'ENSG00000131023', 'HIST1H3B': 'ENSG00000286522', 'TFEB': 'ENSG00000112561', 'ROS1': 'ENSG00000047936', 'PTPRK': 'ENSG00000152894', 'CCNC': 'ENSG00000112237', 'EPHA7': 'ENSG00000135333', 'ECT2L': 'ENSG00000203734', 'CCND3': 'ENSG00000112576', 'POU5F1': 'ENSG00000204531', 'EZR': 'ENSG00000092820', 'ESR1': 'ENSG00000091831', 'PIM1': 'ENSG00000137193', 'IRF4': 'ENSG00000137265', 'FANCE': 'ENSG00000112039', 'AFDN': 'ENSG00000130396', 'HMGA1': 'ENSG00000137309', 'FGFR1OP': 'ENSG00000213066', 'DAXX': 'ENSG00000204209', 'SGK1': 'ENSG00000118515', 'ARID1B': 'ENSG00000049618', 'TNFAIP3': 'ENSG00000118503', 'MYB': 'ENSG00000118513', 'HLA-A': 'ENSG00000206503', 'HIST1H4I': 'ENSG00000276180', 'QKI': 'ENSG00000112531', 'SRSF3': 'ENSG00000112081', 'CDKN1A': 'ENSG00000124762', 'HSP90AB1': 'ENSG00000096384', 'NFKBIE': 'ENSG00000146232', 'SFRP4': 'ENSG00000106483', 'HGF': 'ENSG00000019991', 'SND1': 'ENSG00000197157', 'POT1': 'ENSG00000128513', 'TRIM24': 'ENSG00000122779', 'MACC1': 'ENSG00000183742', 'CDK6': 'ENSG00000105810', 'KIAA1549': 'ENSG00000122778', 'CNTNAP2': 'ENSG00000174469', 'SMO': 'ENSG00000128602', 'CUX1': 'ENSG00000257923', 'FAM131B': 'ENSG00000159784', 'AKAP9': 'ENSG00000127914', 'KMT2C': 'ENSG00000055609', 'FKBP9': 'ENSG00000122642', 'MNX1': 'ENSG00000130675', 'EZH2': 'ENSG00000106462', 'HOXA9': 'ENSG00000078399', 'HOXA11': 'ENSG00000005073', 'TNRC18': 'ENSG00000182095', 'ZNF479': 'ENSG00000185177', 'HIP1': 'ENSG00000127946', 'GTF2I': 'ENSG00000263001', 'JAZF1': 'ENSG00000153814', 'SBDS': 'ENSG00000126524', 'CARD11': 'ENSG00000198286', 'EGFR': 'ENSG00000146648', 'RAC1': 'ENSG00000136238', 'TRRAP': 'ENSG00000196367', 'BRAF': 'ENSG00000157764', 'CREB3L2': 'ENSG00000182158', 'ETV1': 'ENSG00000006468', 'MET': 'ENSG00000105976', 'PMS2': 'ENSG00000122512', 'HNRNPA2B1': 'ENSG00000122566', 'GRM3': 'ENSG00000198822', 'P2RY8': 'ENSG00000182162', 'DCAF12L2': 'ENSG00000198354', 'MTCP1': 'ENSG00000214827', 'ZMYM3': 'ENSG00000147130', 'SSX4': 'ENSG00000268009', 'ATRX': 'ENSG00000085224', 'FLNA': 'ENSG00000196924', 'SMC1A': 'ENSG00000072501', 'NONO': 'ENSG00000147140', 'SSX1': 'ENSG00000126752', 'SEPT6': 'ENSG00000125354', 'BCORL1': 'ENSG00000085185', 'IRS4': 'ENSG00000133124', 'BCOR': 'ENSG00000183337', 'MED12': 'ENSG00000184634', 'PHF6': 'ENSG00000156531', 'ELF4': 'ENSG00000102034', 'GPC3': 'ENSG00000147257', 'SSX2': 'ENSG00000241476', 'RBM10': 'ENSG00000182872', 'ZRSR2': 'ENSG00000169249', 'EIF1AX': 'ENSG00000173674', 'BTK': 'ENSG00000010671', 'FAM47C': 'ENSG00000198173', 'MSN': 'ENSG00000147065', 'TMSB4X': 'ENSG00000205542', 'WAS': 'ENSG00000015285', 'TFE3': 'ENSG00000068323', 'GATA1': 'ENSG00000102145', 'USP9X': 'ENSG00000124486', 'ATP2B3': 'ENSG00000067842', 'CRLF2': 'ENSG00000205755', 'AR': 'ENSG00000169083', 'KDM6A': 'ENSG00000147050', 'STAG2': 'ENSG00000101972', 'DDX3X': 'ENSG00000215301', 'KDM5C': 'ENSG00000126012', 'FOXO4': 'ENSG00000184481', 'LEPROTL1': 'ENSG00000104660', 'LYN': 'ENSG00000254087', 'CDH17': 'ENSG00000079112', 'WRN': 'ENSG00000165392', 'UBR5': 'ENSG00000104517', 'PREX2': 'ENSG00000046889', 'NCOA2': 'ENSG00000140396', 'RSPO2': 'ENSG00000147655', 'CSMD3': 'ENSG00000164796', 'NSD3': 'ENSG00000147548', 'PLAG1': 'ENSG00000181690', 'CHCHD7': 'ENSG00000170791', 'NDRG1': 'ENSG00000104419', 'PABPC1': 'ENSG00000070756', 'EIF3E': 'ENSG00000104408', 'KAT6A': 'ENSG00000083168', 'COX6C': 'ENSG00000164919', 'IKBKB': 'ENSG00000104365', 'EXT1': 'ENSG00000182197', 'ARHGEF10': 'ENSG00000104728', 'RECQL4': 'ENSG00000160957', 'RAD21': 'ENSG00000164754', 'PLEC': 'ENSG00000178209', 'NBN': 'ENSG00000104320', 'CNBD1': 'ENSG00000176571', 'FGFR1': 'ENSG00000077782', 'HOOK3': 'ENSG00000168172', 'PCM1': 'ENSG00000078674', 'ANK1': 'ENSG00000029534', 'HEY1': 'ENSG00000164683', 'MYC': 'ENSG00000136997', 'NRG1': 'ENSG00000157168', 'FAM135B': 'ENSG00000147724', 'RUNX1T1': 'ENSG00000079102', 'TCEA1': 'ENSG00000187735', 'PSIP1': 'ENSG00000164985', 'PAX5': 'ENSG00000196092', 'GNAQ': 'ENSG00000156052', 'KLF4': 'ENSG00000136826', 'RXRA': 'ENSG00000186350', 'MLLT3': 'ENSG00000171843', 'JAK2': 'ENSG00000096968', 'TAL2': 'ENSG00000186051', 'RALGDS': 'ENSG00000160271', 'PPP6C': 'ENSG00000119414', 'OMD': 'ENSG00000127083', 'WNK2': 'ENSG00000165238', 'NTRK2': 'ENSG00000148053', 'SET': 'ENSG00000119335', 'CNTRL': 'ENSG00000119397', 'PTPRD': 'ENSG00000153707', 'NR4A3': 'ENSG00000119508', 'NUP214': 'ENSG00000126883', 'FANCC': 'ENSG00000158169', 'FANCG': 'ENSG00000221829', 'CDKN2A': 'ENSG00000147889', 'CD274': 'ENSG00000120217', 'PDCD1LG2': 'ENSG00000197646', 'FNBP1': 'ENSG00000187239', 'ABL1': 'ENSG00000097007', 'TNC': 'ENSG00000041982', 'NFIB': 'ENSG00000147862', 'PTCH1': 'ENSG00000185920', 'TSC1': 'ENSG00000165699', 'BRD3': 'ENSG00000169925', 'XPA': 'ENSG00000136936', 'SYK': 'ENSG00000165025', 'FANCF': 'ENSG00000183161', 'RRAS2': 'ENSG00000133818', 'ARHGEF12': 'ENSG00000196914', 'KCNJ5': 'ENSG00000120457', 'MUC6': 'ENSG00000184956', 'CREB3L1': 'ENSG00000157613', 'FEN1': 'ENSG00000168496', 'MYOD1': 'ENSG00000129152', 'PAFAH1B2': 'ENSG00000168092', 'CARS': 'ENSG00000110619', 'FAT3': 'ENSG00000165323', 'CLP1': 'ENSG00000172409', 'ZBTB16': 'ENSG00000109906', 'CTNND1': 'ENSG00000198561', 'EXT2': 'ENSG00000151348', 'FADD': 'ENSG00000168040', 'NUMA1': 'ENSG00000137497', 'POU2AF1': 'ENSG00000110777', 'DDX10': 'ENSG00000178105', 'CCND1': 'ENSG00000110092', 'BCL9L': 'ENSG00000186174', 'FOXR1': 'ENSG00000176302', 'LMO1': 'ENSG00000166407', 'FLI1': 'ENSG00000151702', 'DDB2': 'ENSG00000134574', 'CBL': 'ENSG00000110395', 'LMO2': 'ENSG00000135363', 'NUP98': 'ENSG00000110713', 'PICALM': 'ENSG00000073921', 'KMT2A': 'ENSG00000118058', 'YAP1': 'ENSG00000137693', 'BIRC3': 'ENSG00000023445', 'HRAS': 'ENSG00000174775', 'ATM': 'ENSG00000149311', 'SDHAF2': 'ENSG00000167985', 'SDHD': 'ENSG00000204370', 'DDX6': 'ENSG00000110367', 'MAML2': 'ENSG00000184384', 'EED': 'ENSG00000074266', 'WT1': 'ENSG00000184937', 'KLF6': 'ENSG00000067082', 'KIF5B': 'ENSG00000170759', 'KAT6B': 'ENSG00000156650', 'CPEB3': 'ENSG00000107864', 'MGMT': 'ENSG00000170430', 'SHTN1': 'ENSG00000187164', 'SUFU': 'ENSG00000107882', 'MLLT10': 'ENSG00000078403', 'NUTM2D': 'ENSG00000214562', 'CYP2C8': 'ENSG00000138115', 'PRF1': 'ENSG00000180644', 'TLX1': 'ENSG00000107807', 'FAS': 'ENSG00000026103', 'NT5C2': 'ENSG00000076685', 'NFKB2': 'ENSG00000077150', 'NUTM2B': 'ENSG00000188199', 'CCDC6': 'ENSG00000108091', 'GATA3': 'ENSG00000107485', 'TCF7L2': 'ENSG00000148737', 'VTI1A': 'ENSG00000151532', 'A1CF': 'ENSG00000148584', 'ABI1': 'ENSG00000136754', 'FGFR2': 'ENSG00000066468', 'TET1': 'ENSG00000138336', 'BMPR1A': 'ENSG00000107779', 'PTEN': 'ENSG00000171862', 'RET': 'ENSG00000165731', 'ETV6': 'ENSG00000139083', 'BTG1': 'ENSG00000133639', 'CHST11': 'ENSG00000171310', 'SH2B3': 'ENSG00000111252', 'RAP1B': 'ENSG00000127314', 'USP44': 'ENSG00000136014', 'TBX3': 'ENSG00000135111', 'GLI1': 'ENSG00000111087', 'CDK4': 'ENSG00000135446', 'PTPN6': 'ENSG00000111679', 'CCND2': 'ENSG00000118971', 'NAB2': 'ENSG00000166886', 'COL2A1': 'ENSG00000139219', 'PPFIBP1': 'ENSG00000110841', 'ACVR1B': 'ENSG00000135503', 'NACA': 'ENSG00000196531', 'CHD4': 'ENSG00000111642', 'HMGA2': 'ENSG00000149948', 'PTPRB': 'ENSG00000127329', 'WIF1': 'ENSG00000156076', 'CLIP1': 'ENSG00000130779', 'ERBB3': 'ENSG00000065361', 'CDKN1B': 'ENSG00000111276', 'DDIT3': 'ENSG00000175197', 'KMT2D': 'ENSG00000167548', 'PTPN11': 'ENSG00000179295', 'MDM2': 'ENSG00000135679', 'ERC1': 'ENSG00000082805', 'ATF1': 'ENSG00000123268', 'PRPF40B': 'ENSG00000110844', 'HOXC13': 'ENSG00000123364', 'HOXC11': 'ENSG00000123388', 'POLE': 'ENSG00000177084', 'KDM5A': 'ENSG00000073614', 'NCOR2': 'ENSG00000196498', 'HNF1A': 'ENSG00000135100', 'ZNF384': 'ENSG00000126746', 'KRAS': 'ENSG00000133703', 'SMARCD1': 'ENSG00000066117', 'SETD1B': 'ENSG00000139718', 'BCL7A': 'ENSG00000110987', 'ZCCHC8': 'ENSG00000033030', 'ALDH2': 'ENSG00000111275', 'ARID2': 'ENSG00000189079', 'STAT6': 'ENSG00000166888', 'ETNK1': 'ENSG00000139163', 'LRIG3': 'ENSG00000139263', 'LHFPL6': 'ENSG00000183722', 'FOXO1': 'ENSG00000150907', 'FLT3': 'ENSG00000122025', 'LCP1': 'ENSG00000136167', 'ZMYM2': 'ENSG00000121741', 'LATS2': 'ENSG00000150457', 'CYSLTR2': 'ENSG00000152207', 'NBEA': 'ENSG00000172915', 'SOX21': 'ENSG00000125285', 'GPC5': 'ENSG00000179399', 'CDX2': 'ENSG00000165556', 'BRCA2': 'ENSG00000139618', 'RB1': 'ENSG00000139687', 'GOLGA5': 'ENSG00000066455', 'KTN1': 'ENSG00000126777', 'TCL1A': 'ENSG00000100721', 'GPHN': 'ENSG00000171723', 'TSHR': 'ENSG00000165409', 'TRIP11': 'ENSG00000100815', 'HIF1A': 'ENSG00000100644', 'NKX2-1': 'ENSG00000136352', 'PRKD1': 'ENSG00000184304', 'ARHGAP5': 'ENSG00000100852', 'BAZ1A': 'ENSG00000198604', 'HSP90AA1': 'ENSG00000080824', 'FOXA1': 'ENSG00000129514', 'DICER1': 'ENSG00000100697', 'BCL11B': 'ENSG00000127152', 'NIN': 'ENSG00000100503', 'MAX': 'ENSG00000125952', 'RAD51B': 'ENSG00000182185', 'AKT1': 'ENSG00000142208', 'PML': 'ENSG00000140464', 'NR2F2': 'ENSG00000185551', 'BUB1B': 'ENSG00000156970', 'TCF12': 'ENSG00000140262', 'USP8': 'ENSG00000138592', 'FES': 'ENSG00000182511', 'BLM': 'ENSG00000197299', 'IDH2': 'ENSG00000182054', 'MAP2K1': 'ENSG00000169032', 'MYO5A': 'ENSG00000197535', 'KNSTRN': 'ENSG00000128944', 'CRTC3': 'ENSG00000140577', 'KNL1': 'ENSG00000137812', 'CHD2': 'ENSG00000173575', 'POLG': 'ENSG00000140521', 'C15orf65': 'ENSG00000261652', 'NUTM1': 'ENSG00000184507', 'SMAD3': 'ENSG00000166949', 'AXIN1': 'ENSG00000103126', 'CREBBP': 'ENSG00000005339', 'SNX29': 'ENSG00000048471', 'MAF': 'ENSG00000178573', 'CDH11': 'ENSG00000140937', 'ZFHX3': 'ENSG00000140836', 'RFWD3': 'ENSG00000168411', 'RMI2': 'ENSG00000175643', 'CYLD': 'ENSG00000083799', 'HERPUD1': 'ENSG00000051108', 'CDH1': 'ENSG00000039068', 'ERCC4': 'ENSG00000175595', 'PRKCB': 'ENSG00000166501', 'CBFB': 'ENSG00000067955', 'SOCS1': 'ENSG00000185338', 'CIITA': 'ENSG00000179583', 'CTCF': 'ENSG00000102974', 'CBFA2T3': 'ENSG00000129993', 'FANCA': 'ENSG00000187741', 'IL21R': 'ENSG00000103522', 'TNFRSF17': 'ENSG00000048462', 'TRAF7': 'ENSG00000131653', 'MYH11': 'ENSG00000133392', 'PALB2': 'ENSG00000083093', 'TSC2': 'ENSG00000103197', 'NTHL1': 'ENSG00000065057', 'FUS': 'ENSG00000089280', 'GRIN2A': 'ENSG00000183454', 'SPOP': 'ENSG00000121067', 'FLCN': 'ENSG00000154803', 'MAP2K4': 'ENSG00000065559', 'H3F3B': 'ENSG00000132475', 'MSI2': 'ENSG00000153944', 'DDX5': 'ENSG00000108654', 'POLR2A': 'ENSG00000181222', 'GAS7': 'ENSG00000007237', 'SEPT9': 'ENSG00000184640', 'YWHAE': 'ENSG00000108953', 'COL1A1': 'ENSG00000108821', 'BRIP1': 'ENSG00000136492', 'CANT1': 'ENSG00000171302', 'PPM1D': 'ENSG00000170836', 'SPECC1': 'ENSG00000128487', 'USP6': 'ENSG00000129204', 'RAD51C': 'ENSG00000108384', 'CD79B': 'ENSG00000007312', 'STAT5B': 'ENSG00000173757', 'RNF43': 'ENSG00000108375', 'RABEP1': 'ENSG00000029725', 'NCOR1': 'ENSG00000141027', 'RNF213': 'ENSG00000173821', 'ETV4': 'ENSG00000175832', 'AXIN2': 'ENSG00000168646', 'HLF': 'ENSG00000108924', 'NF1': 'ENSG00000196712', 'KAT7': 'ENSG00000136504', 'CDK12': 'ENSG00000167258', 'PRKAR1A': 'ENSG00000108946', 'SRSF2': 'ENSG00000161547', 'TP53': 'ENSG00000141510', 'STAT3': 'ENSG00000168610', 'BRCA1': 'ENSG00000012048', 'LASP1': 'ENSG00000002834', 'SUZ12': 'ENSG00000178691', 'ASPSCR1': 'ENSG00000169696', 'PER1': 'ENSG00000179094', 'ERBB2': 'ENSG00000141736', 'IKZF3': 'ENSG00000161405', 'RARA': 'ENSG00000131759', 'CCR7': 'ENSG00000126353', 'SMARCE1': 'ENSG00000073584', 'SS18': 'ENSG00000141380', 'SMAD2': 'ENSG00000175387', 'ZNF521': 'ENSG00000198795', 'BCL2': 'ENSG00000171791', 'DCC': 'ENSG00000187323', 'SMAD4': 'ENSG00000141646', 'SALL4': 'ENSG00000101115', 'SS18L1': 'ENSG00000184402', 'NFATC2': 'ENSG00000101096', 'SDC4': 'ENSG00000124145', 'PTK6': 'ENSG00000101213', 'TOP1': 'ENSG00000198900', 'MAFB': 'ENSG00000204103', 'CRNKL1': 'ENSG00000101343', 'PLCG1': 'ENSG00000124181', 'SRC': 'ENSG00000197122', 'PTPRT': 'ENSG00000196090', 'GNAS': 'ENSG00000087460', 'ASXL1': 'ENSG00000171456', 'SIRPA': 'ENSG00000198053', 'VAV1': 'ENSG00000141968', 'GNA11': 'ENSG00000088256', 'CRTC1': 'ENSG00000105662', 'CCNE1': 'ENSG00000105173', 'KLK2': 'ENSG00000167751', 'CEP89': 'ENSG00000121289', 'LYL1': 'ENSG00000104903', 'SH3GL1': 'ENSG00000141985', 'CD209': 'ENSG00000090659', 'KEAP1': 'ENSG00000079999', 'ZNF331': 'ENSG00000130844', 'BRD4': 'ENSG00000141867', 'CALR': 'ENSG00000179218', 'LSM14A': 'ENSG00000257103', 'TFPT': 'ENSG00000105619', 'ERCC2': 'ENSG00000104884', 'CIC': 'ENSG00000079432', 'MAP2K2': 'ENSG00000126934', 'POLD1': 'ENSG00000062822', 'ZNF429': 'ENSG00000197013', 'BAX': 'ENSG00000087088', 'FSTL3': 'ENSG00000070404', 'ELL': 'ENSG00000105656', 'JAK3': 'ENSG00000105639', 'CD79A': 'ENSG00000105369', 'PRKACA': 'ENSG00000072062', 'AKT2': 'ENSG00000105221', 'CNOT3': 'ENSG00000088038', 'PPP2R1A': 'ENSG00000105568', 'TCF3': 'ENSG00000071564', 'CEBPA': 'ENSG00000245848', 'BCL3': 'ENSG00000069399', 'DNM2': 'ENSG00000079805', 'SMARCA4': 'ENSG00000127616', 'MUC16': 'ENSG00000181143', 'MLLT1': 'ENSG00000130382', 'STK11': 'ENSG00000118046', 'DNAJB1': 'ENSG00000132002', 'TPM4': 'ENSG00000167460', 'PIK3R2': 'ENSG00000105647', 'ISX': 'ENSG00000175329', 'PDGFB': 'ENSG00000100311', 'ZNRF3': 'ENSG00000183579', 'MAPK1': 'ENSG00000100030', 'APOBEC3B': 'ENSG00000179750', 'SEPT5': 'ENSG00000184702', 'BCR': 'ENSG00000186716', 'MYH9': 'ENSG00000100345', 'MN1': 'ENSG00000169184', 'SMARCB1': 'ENSG00000099956', 'EWSR1': 'ENSG00000182944', 'CLTCL1': 'ENSG00000070371', 'DGCR8': 'ENSG00000128191', 'CHEK2': 'ENSG00000183765', 'NF2': 'ENSG00000186575', 'PATZ1': 'ENSG00000100105', 'EP300': 'ENSG00000100393', 'U2AF1': 'ENSG00000160201', 'TMPRSS2': 'ENSG00000184012', 'OLIG2': 'ENSG00000205927', 'ERG': 'ENSG00000157554', 'RUNX1': 'ENSG00000159216'}\n"
     ]
    }
   ],
   "source": [
    "# RELIER LE NOM USUEL DU GÈNE (EX. BRCA1) UTILISÉ SUR COSMICS À SON TRANSCRIT DE RÉFÉRENCE\n",
    "les_transcrits_de_reference = {}\n",
    "import re\n",
    "data = '../data/Cosmic_MutantCensus_v101_GRCh38.tsv' # Catalogue COSMICS: contient toutes les mutations de tous les gènes\n",
    "with open(data, 'r') as file: \n",
    "    next(file) # Skip la première ligne (header)\n",
    "    for line in file:\n",
    "        colonne = line.split('\\t')\n",
    "        gene_name = colonne[0] # Nom usuel sur COSMICS (ex. BRCA1)\n",
    "        transcrit_name_info = re.search(r'ENST[0123456789]{5,100}', colonne[2])\n",
    "        transcrit_reference = transcrit_name_info.group(0) # Transcrit de référence, les mutations sont établies selon ce transcrit\n",
    "        les_transcrits_de_reference[gene_name]= transcrit_reference # Clé = nom usuel du gène, Valeur = son transcrit de référence\n",
    "\n",
    "# RELIER LE NOM USUEL DU GÈNE (EX. BRCA1) UTILISÉ SUR COSMICS À SON IDENTIFIANT ENSG     \n",
    "equivalence_nom = {}\n",
    "with open('../data/Homo_sapiens.GRCh38.113.gtf', 'r') as file: # GTF\n",
    "    for line in file:\n",
    "        colonne_tab = line.split('\\t')\n",
    "        colonne_guillemet = line.split('\"')\n",
    "        if len(colonne_tab) > 3: # Pour enlever les trois premières lignes du file qui n'ont pas rapport\n",
    "            for valeur in les_transcrits_de_reference.values(): # Itération sur le transcrit de référence pour tous les gènes\n",
    "                if colonne_tab[2] == 'transcript' and valeur in line: # Sélectionne la ligne du transcrit de référence\n",
    "                    gene = colonne_guillemet[1] # Identifiant ENSG du gène\n",
    "                    list_gene_name = [cle for cle, v in les_transcrits_de_reference.items() if v == valeur] # Je recherche le nom usuel du gène pour lequel je viens de trouver le transcrit de référence\n",
    "                    cosmic_gene_name = list_gene_name[0]\n",
    "                    equivalence_nom[cosmic_gene_name]= gene # Établit un lien entre le nom usuel et ENSG\n",
    "#print(equivalence_nom)\n",
    "\n",
    "# Il y a 35 gènes/741 gènes présents dans la database de Cosmic dont le transcrit de référence ne se retrouve pas dans le GTF. \n",
    "# COSMICS utilise uniquement le nom usuel du gène dans sa database et le GTF utilise l'identifiant ENSG, pour l'instant, mon code établit le lien\n",
    "# nom usuel-identifiant ENSG grâce à l'identifiant ENST du transcrit de référence.\n",
    "# Ces gènes ne pourront pas être effectué pour le moment, car Cosmics donne ses positions de mutations par rapport au transcrit de référence.\n",
    "# Il reste 706 gènes qui sont traités. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "527f2385",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ENSG00000142611': '+', 'ENSG00000054118': '+', 'ENSG00000123080': '+', 'ENSG00000183508': '+', 'ENSG00000047410': '-', 'ENSG00000116132': '+', 'ENSG00000074964': '+', 'ENSG00000117318': '-'}\n",
      "{'PRDM16': '+', 'THRAP3': '+', 'CDKN2C': '+', 'TENT5C': '+', 'TPR': '-', 'PRRX1': '+', 'ARHGEF10L': '+', 'ID3': '-'}\n",
      "{'ENSG00000142611': {'ENST00000511072': {'ENSE00002048533': ['3069168', '3069296'], 'ENSE00001754112': ['3186125', '3186474'], 'ENSE00003480863': ['3244087', '3244137'], 'ENSE00002034212': ['3385149', '3385286'], 'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00003701451': ['3414560', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00002081080': ['3433677', '3434342']}, 'ENST00000607632': {'ENSE00003700259': ['3069183', '3069296'], 'ENSE00003695128': ['3186125', '3186591']}, 'ENST00000378391': {'ENSE00001222906': ['3069197', '3069296'], 'ENSE00001754112': ['3186125', '3186474'], 'ENSE00003480863': ['3244087', '3244137'], 'ENSE00003695725': ['3385152', '3385286'], 'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00003701451': ['3414560', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00003700833': ['3431966', '3432140'], 'ENSE00001477303': ['3433734', '3435421']}, 'ENST00000514189': {'ENSE00001598723': ['3069211', '3069296'], 'ENSE00001754112': ['3186125', '3186474'], 'ENSE00003480863': ['3244087', '3244137'], 'ENSE00002034212': ['3385149', '3385286'], 'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00001582537': ['3414563', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00002081080': ['3433677', '3434342']}, 'ENST00000270722': {'ENSE00003850248': ['3069203', '3069296'], 'ENSE00001754112': ['3186125', '3186474'], 'ENSE00003480863': ['3244087', '3244137'], 'ENSE00003695725': ['3385152', '3385286'], 'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00003701451': ['3414560', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00003700833': ['3431966', '3432140'], 'ENSE00001759023': ['3433677', '3438621']}, 'ENST00000512462': {'ENSE00002087812': ['3237931', '3238095'], 'ENSE00003516886': ['3244087', '3244137'], 'ENSE00003662380': ['3385152', '3385286'], 'ENSE00003697266': ['3396491', '3396593'], 'ENSE00003701239': ['3402791', '3402998'], 'ENSE00003700436': ['3404739', '3404886'], 'ENSE00003695760': ['3405495', '3405648'], 'ENSE00003699089': ['3411384', '3412800'], 'ENSE00003702178': ['3414560', '3414647'], 'ENSE00003701293': ['3417828', '3417997'], 'ENSE00003697785': ['3418667', '3418744'], 'ENSE00003701636': ['3425581', '3425750'], 'ENSE00003696062': ['3426051', '3426225'], 'ENSE00003697832': ['3430872', '3431108'], 'ENSE00003698886': ['3431966', '3432140'], 'ENSE00002068021': ['3433677', '3433925']}, 'ENST00000463591': {'ENSE00001882136': ['3244132', '3244137'], 'ENSE00003695725': ['3385152', '3385286'], 'ENSE00001846205': ['3386676', '3386918']}, 'ENST00000509860': {'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00001582537': ['3414563', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00003700833': ['3431966', '3432140'], 'ENSE00001477317': ['3433677', '3434293']}, 'ENST00000378389': {'ENSE00001905893': ['3424919', '3425059'], 'ENSE00003701636': ['3425581', '3425750'], 'ENSE00003696062': ['3426051', '3426225'], 'ENSE00003697832': ['3430872', '3431108'], 'ENSE00001875784': ['3433677', '3434095']}, 'ENST00000606170': {'ENSE00003697875': ['3425216', '3425750'], 'ENSE00003700534': ['3426051', '3426072']}}, 'ENSG00000054118': {'ENST00000354618': {'ENSE00001459618': ['36224443', '36224505'], 'ENSE00001354450': ['36259382', '36259484'], 'ENSE00001298923': ['36282533', '36282700'], 'ENSE00001291272': ['36286368', '36287270'], 'ENSE00000955947': ['36289060', '36289764'], 'ENSE00003566879': ['36291374', '36291546'], 'ENSE00003479354': ['36292598', '36292709'], 'ENSE00003660836': ['36293851', '36293935'], 'ENSE00000766231': ['36296583', '36296770'], 'ENSE00001079145': ['36300886', '36301084'], 'ENSE00001079144': ['36301553', '36301696'], 'ENSE00001402345': ['36303796', '36305357']}, 'ENST00000469141': {'ENSE00001829811': ['36224432', '36224505'], 'ENSE00001910590': ['36224971', '36225102'], 'ENSE00001354450': ['36259382', '36259484'], 'ENSE00001298923': ['36282533', '36282700'], 'ENSE00001291272': ['36286368', '36287270'], 'ENSE00000955947': ['36289060', '36289764'], 'ENSE00003566879': ['36291374', '36291546'], 'ENSE00003479354': ['36292598', '36292709'], 'ENSE00003660836': ['36293851', '36293935'], 'ENSE00000766231': ['36296583', '36296770'], 'ENSE00001079145': ['36300886', '36301084'], 'ENSE00001079144': ['36301553', '36301696'], 'ENSE00002179918': ['36303796', '36304148']}, 'ENST00000648638': {'ENSE00001861135': ['36224433', '36224505'], 'ENSE00001917968': ['36225412', '36225463'], 'ENSE00001354450': ['36259382', '36259484'], 'ENSE00001298923': ['36282533', '36282700'], 'ENSE00001291272': ['36286368', '36287270'], 'ENSE00000955947': ['36289060', '36289764'], 'ENSE00003566879': ['36291374', '36291546'], 'ENSE00003479354': ['36292598', '36292709'], 'ENSE00003660836': ['36293851', '36293935'], 'ENSE00000766231': ['36296583', '36296770'], 'ENSE00001079145': ['36300886', '36301084'], 'ENSE00001079144': ['36301553', '36301696'], 'ENSE00003840935': ['36303796', '36303831'], 'ENSE00003839481': ['36304474', '36305357']}, 'ENST00000466743': {'ENSE00001941050': ['36289718', '36289764'], 'ENSE00003588907': ['36291374', '36291546'], 'ENSE00003608584': ['36292598', '36292709'], 'ENSE00003570214': ['36293851', '36293935'], 'ENSE00001859170': ['36294050', '36294169']}}, 'ENSG00000123080': {'ENST00000262662': {'ENSE00001248960': ['50960745', '50960803'], 'ENSE00001248955': ['50967936', '50969899'], 'ENSE00001064516': ['50970358', '50970497'], 'ENSE00001064515': ['50973893', '50974633']}, 'ENST00000396148': {'ENSE00001524054': ['50968706', '50969899'], 'ENSE00001064516': ['50970358', '50970497'], 'ENSE00001456048': ['50973893', '50974634']}, 'ENST00000371761': {'ENSE00001456052': ['50970245', '50970497'], 'ENSE00001456048': ['50973893', '50974634']}}, 'ENSG00000183508': {'ENST00000369448': {'ENSE00001450051': ['117606048', '117606153'], 'ENSE00001450050': ['117622842', '117628389']}}, 'ENSG00000047410': {'ENST00000367478': {'ENSE00003843206': ['186374878', '186375253'], 'ENSE00003487734': ['186373359', '186373463'], 'ENSE00003587929': ['186370970', '186371043'], 'ENSE00003662167': ['186367886', '186367982'], 'ENSE00003522190': ['186363342', '186363445'], 'ENSE00003577455': ['186362837', '186363001'], 'ENSE00003490606': ['186362288', '186362380'], 'ENSE00003604606': ['186361789', '186361869'], 'ENSE00003514225': ['186361622', '186361709'], 'ENSE00003535023': ['186360765', '186360905'], 'ENSE00003689727': ['186360273', '186360364'], 'ENSE00003477153': ['186359799', '186359996'], 'ENSE00003645437': ['186358543', '186358650'], 'ENSE00003608274': ['186357397', '186357623'], 'ENSE00003685897': ['186356286', '186356449'], 'ENSE00003582311': ['186355635', '186355768'], 'ENSE00003610313': ['186355410', '186355558'], 'ENSE00003670368': ['186353688', '186353850'], 'ENSE00003582315': ['186351976', '186352110'], 'ENSE00001030868': ['186351330', '186351470'], 'ENSE00001030871': ['186350223', '186350388'], 'ENSE00001030911': ['186347292', '186347458'], 'ENSE00001030867': ['186346135', '186346287'], 'ENSE00001068225': ['186345580', '186345696'], 'ENSE00001068221': ['186344375', '186344578'], 'ENSE00003529386': ['186343906', '186344090'], 'ENSE00001265722': ['186343326', '186343473'], 'ENSE00001068222': ['186341252', '186341389'], 'ENSE00003623500': ['186341028', '186341159'], 'ENSE00003615941': ['186339642', '186339772'], 'ENSE00001030928': ['186338033', '186338243'], 'ENSE00001030892': ['186337013', '186337156'], 'ENSE00001030903': ['186336496', '186336694'], 'ENSE00001030889': ['186335338', '186335543'], 'ENSE00001030922': ['186335068', '186335129'], 'ENSE00001030902': ['186334325', '186334533'], 'ENSE00001030886': ['186333122', '186333394'], 'ENSE00001030900': ['186332195', '186332343'], 'ENSE00001030927': ['186331498', '186331581'], 'ENSE00001444649': ['186327460', '186327660'], 'ENSE00001030908': ['186326104', '186326235'], 'ENSE00000790831': ['186325764', '186325854'], 'ENSE00000790830': ['186323686', '186323870'], 'ENSE00000790829': ['186322518', '186322586'], 'ENSE00003507780': ['186322318', '186322412'], 'ENSE00003525673': ['186320312', '186320418'], 'ENSE00003691190': ['186318733', '186318828'], 'ENSE00000790825': ['186318447', '186318603'], 'ENSE00000790824': ['186317482', '186317600'], 'ENSE00000823022': ['186314629', '186314724'], 'ENSE00001444646': ['186311652', '186314026']}, 'ENST00000467810': {'ENSE00001945374': ['186322518', '186322701'], 'ENSE00003499052': ['186322318', '186322412'], 'ENSE00003629978': ['186320312', '186320418'], 'ENSE00003562529': ['186318733', '186318828'], 'ENSE00001817070': ['186318505', '186318603']}, 'ENST00000492973': {'ENSE00001871801': ['186326104', '186326220'], 'ENSE00001887741': ['186325519', '186325854']}, 'ENST00000481347': {'ENSE00003655025': ['186343906', '186344090'], 'ENSE00001945728': ['186341252', '186343473'], 'ENSE00003528301': ['186341028', '186341159'], 'ENSE00003549857': ['186339642', '186339772'], 'ENSE00001842188': ['186338187', '186338243']}, 'ENST00000491783': {'ENSE00001901888': ['186356286', '186356538'], 'ENSE00003466865': ['186355635', '186355768'], 'ENSE00003471722': ['186355410', '186355558'], 'ENSE00003505085': ['186353688', '186353850'], 'ENSE00003511424': ['186351976', '186352110'], 'ENSE00001934432': ['186351423', '186351470']}, 'ENST00000469463': {'ENSE00001948608': ['186353688', '186353948'], 'ENSE00001831515': ['186351952', '186352110']}, 'ENST00000474852': {'ENSE00003677161': ['186374878', '186375325'], 'ENSE00003490929': ['186373359', '186373463'], 'ENSE00003620212': ['186370970', '186371043'], 'ENSE00003545532': ['186367886', '186367982'], 'ENSE00003611427': ['186363342', '186363445'], 'ENSE00003651485': ['186362837', '186363001'], 'ENSE00003638079': ['186362288', '186362380'], 'ENSE00003527111': ['186361789', '186361869'], 'ENSE00003517331': ['186361622', '186361709'], 'ENSE00003594539': ['186360765', '186360905'], 'ENSE00003693781': ['186360273', '186360364'], 'ENSE00003689787': ['186359799', '186359996'], 'ENSE00003619958': ['186358543', '186358650'], 'ENSE00003478760': ['186357397', '186357623'], 'ENSE00003662843': ['186356286', '186356449'], 'ENSE00003466865': ['186355635', '186355768'], 'ENSE00003471722': ['186355410', '186355558'], 'ENSE00001893692': ['186353688', '186353880'], 'ENSE00001896671': ['186352015', '186352110']}, 'ENST00000451586': {'ENSE00001617725': ['186375559', '186375693'], 'ENSE00001068219': ['186374878', '186375186'], 'ENSE00001753818': ['186373393', '186373463']}, 'ENST00000613151': {'ENSE00003466314': ['186374878', '186375325'], 'ENSE00003487734': ['186373359', '186373463'], 'ENSE00003587929': ['186370970', '186371043'], 'ENSE00003662167': ['186367886', '186367982'], 'ENSE00003522190': ['186363342', '186363445'], 'ENSE00003577455': ['186362837', '186363001'], 'ENSE00003490606': ['186362288', '186362380'], 'ENSE00003604606': ['186361789', '186361869'], 'ENSE00003514225': ['186361622', '186361709'], 'ENSE00003535023': ['186360765', '186360905'], 'ENSE00003689727': ['186360273', '186360364'], 'ENSE00003477153': ['186359799', '186359996'], 'ENSE00003645437': ['186358543', '186358650'], 'ENSE00003608274': ['186357397', '186357623'], 'ENSE00003685897': ['186356286', '186356449'], 'ENSE00003582311': ['186355635', '186355768'], 'ENSE00003610313': ['186355410', '186355558'], 'ENSE00003733373': ['186353871', '186353880']}}, 'ENSG00000116132': {'ENST00000553786': {'ENSE00002433506': ['170662728', '170662924'], 'ENSE00001814256': ['170664306', '170664459'], 'ENSE00002386431': ['170719726', '170719903']}, 'ENST00000367760': {'ENSE00001907015': ['170663134', '170664459'], 'ENSE00002347336': ['170719726', '170719901'], 'ENSE00003617120': ['170726220', '170726401'], 'ENSE00003697545': ['170730277', '170730348'], 'ENSE00001834707': ['170736048', '170736508']}, 'ENST00000239461': {'ENSE00001445579': ['170664131', '170664459'], 'ENSE00002347336': ['170719726', '170719901'], 'ENSE00003617120': ['170726220', '170726401'], 'ENSE00001821799': ['170736048', '170739421']}, 'ENST00000497230': {'ENSE00002447784': ['170663937', '170664459'], 'ENSE00002347336': ['170719726', '170719901'], 'ENSE00001919426': ['170726220', '170729021']}, 'ENST00000476867': {'ENSE00001873292': ['170722617', '170722668'], 'ENSE00003530720': ['170726220', '170726401'], 'ENSE00002517633': ['170736048', '170736399']}, 'ENST00000495280': {'ENSE00001936632': ['170722645', '170722668'], 'ENSE00003530720': ['170726220', '170726401'], 'ENSE00003697845': ['170730277', '170730348'], 'ENSE00001913921': ['170736048', '170736233']}, 'ENST00000485529': {'ENSE00001889459': ['170730206', '170730348'], 'ENSE00001913921': ['170736048', '170736233']}, 'ENST00000496573': {'ENSE00001941384': ['170730446', '170730621'], 'ENSE00001913921': ['170736048', '170736233']}}, 'ENSG00000074964': {'ENST00000361221': {'ENSE00001267745': ['17539698', '17539950'], 'ENSE00001162086': ['17580553', '17580632'], 'ENSE00001161931': ['17587460', '17587645'], 'ENSE00001182858': ['17588446', '17588479'], 'ENSE00003649194': ['17602127', '17602218'], 'ENSE00003548497': ['17603508', '17603591'], 'ENSE00003670661': ['17607802', '17607977'], 'ENSE00000896633': ['17613058', '17613174'], 'ENSE00000896634': ['17616094', '17616202'], 'ENSE00003503885': ['17619339', '17619445'], 'ENSE00003505774': ['17621864', '17621941'], 'ENSE00003599477': ['17622996', '17623175'], 'ENSE00003493681': ['17624387', '17624503'], 'ENSE00003641913': ['17625956', '17626048'], 'ENSE00003583343': ['17627330', '17627503'], 'ENSE00003654161': ['17632321', '17632466'], 'ENSE00001755595': ['17634548', '17634562'], 'ENSE00003461835': ['17634835', '17635016'], 'ENSE00003650063': ['17637888', '17638003'], 'ENSE00003685080': ['17638562', '17638689'], 'ENSE00003669483': ['17640202', '17640302'], 'ENSE00003631164': ['17648554', '17648675'], 'ENSE00003678534': ['17654636', '17654722'], 'ENSE00003507371': ['17655879', '17656102'], 'ENSE00003468102': ['17656554', '17656708'], 'ENSE00003607478': ['17664447', '17664595'], 'ENSE00003506428': ['17687573', '17687747'], 'ENSE00003460282': ['17695158', '17695280'], 'ENSE00001954969': ['17696848', '17697869']}, 'ENST00000375415': {'ENSE00001162086': ['17580553', '17580632'], 'ENSE00001161931': ['17587460', '17587645'], 'ENSE00001182858': ['17588446', '17588479'], 'ENSE00003649194': ['17602127', '17602218'], 'ENSE00003548497': ['17603508', '17603591'], 'ENSE00003670661': ['17607802', '17607977'], 'ENSE00000896634': ['17616094', '17616202'], 'ENSE00003503885': ['17619339', '17619445'], 'ENSE00003505774': ['17621864', '17621941'], 'ENSE00003599477': ['17622996', '17623175'], 'ENSE00003493681': ['17624387', '17624503'], 'ENSE00003641913': ['17625956', '17626048'], 'ENSE00003583343': ['17627330', '17627503'], 'ENSE00003654161': ['17632321', '17632466'], 'ENSE00001755595': ['17634548', '17634562'], 'ENSE00003461835': ['17634835', '17635016'], 'ENSE00003650063': ['17637888', '17638003'], 'ENSE00003685080': ['17638562', '17638689'], 'ENSE00003669483': ['17640202', '17640302'], 'ENSE00003631164': ['17648554', '17648675'], 'ENSE00003678534': ['17654636', '17654722'], 'ENSE00003507371': ['17655879', '17656102'], 'ENSE00003468102': ['17656554', '17656708'], 'ENSE00003607478': ['17664447', '17664595'], 'ENSE00003506428': ['17687573', '17687747'], 'ENSE00003460282': ['17695158', '17695280'], 'ENSE00001467009': ['17696848', '17697649']}, 'ENST00000482892': {'ENSE00001850961': ['17602171', '17602218'], 'ENSE00003623586': ['17603508', '17603591'], 'ENSE00001932431': ['17604476', '17604882']}, 'ENST00000469726': {'ENSE00001855212': ['17615088', '17616202'], 'ENSE00003616000': ['17619339', '17619445'], 'ENSE00003668894': ['17621864', '17621941'], 'ENSE00003594571': ['17622996', '17623175'], 'ENSE00003555762': ['17624387', '17624503'], 'ENSE00003659199': ['17625956', '17626048'], 'ENSE00003658233': ['17627330', '17627503'], 'ENSE00003686475': ['17632321', '17632466'], 'ENSE00003558496': ['17634835', '17635016'], 'ENSE00003602420': ['17637888', '17638003'], 'ENSE00003576297': ['17638562', '17638689'], 'ENSE00003522686': ['17640202', '17640302'], 'ENSE00003486711': ['17648554', '17648675'], 'ENSE00003515029': ['17654636', '17654722'], 'ENSE00003685692': ['17655879', '17656102'], 'ENSE00003466542': ['17656554', '17656708'], 'ENSE00003468557': ['17664447', '17664595'], 'ENSE00003526832': ['17687573', '17687747'], 'ENSE00003499717': ['17695158', '17695280'], 'ENSE00001889352': ['17696848', '17697860']}, 'ENST00000375408': {'ENSE00001935881': ['17618316', '17618522'], 'ENSE00003503885': ['17619339', '17619445'], 'ENSE00003505774': ['17621864', '17621941'], 'ENSE00003599477': ['17622996', '17623175'], 'ENSE00003493681': ['17624387', '17624503'], 'ENSE00003641913': ['17625956', '17626048'], 'ENSE00003583343': ['17627330', '17627503'], 'ENSE00003654161': ['17632321', '17632466'], 'ENSE00003461835': ['17634835', '17635016'], 'ENSE00003650063': ['17637888', '17638003'], 'ENSE00003685080': ['17638562', '17638689'], 'ENSE00003669483': ['17640202', '17640302'], 'ENSE00003631164': ['17648554', '17648675'], 'ENSE00003678534': ['17654636', '17654722'], 'ENSE00003507371': ['17655879', '17656102'], 'ENSE00003468102': ['17656554', '17656708'], 'ENSE00003607478': ['17664447', '17664595'], 'ENSE00003506428': ['17687573', '17687747'], 'ENSE00003460282': ['17695158', '17695280'], 'ENSE00003544025': ['17696848', '17697868']}, 'ENST00000167825': {'ENSE00001833298': ['17626007', '17626048'], 'ENSE00003583343': ['17627330', '17627503'], 'ENSE00003654161': ['17632321', '17632466'], 'ENSE00003461835': ['17634835', '17635016'], 'ENSE00003650063': ['17637888', '17638003'], 'ENSE00003685080': ['17638562', '17638689'], 'ENSE00003669483': ['17640202', '17640302'], 'ENSE00003631164': ['17648554', '17648675'], 'ENSE00003678534': ['17654636', '17654722'], 'ENSE00003507371': ['17655879', '17656102'], 'ENSE00003468102': ['17656554', '17656708'], 'ENSE00003607478': ['17664447', '17664595'], 'ENSE00003506428': ['17687573', '17687747'], 'ENSE00001952719': ['17691120', '17691192'], 'ENSE00003499717': ['17695158', '17695280'], 'ENSE00003546919': ['17696848', '17697868']}, 'ENST00000482359': {'ENSE00001820929': ['17634257', '17634562'], 'ENSE00003558496': ['17634835', '17635016'], 'ENSE00003602420': ['17637888', '17638003']}, 'ENST00000466782': {'ENSE00001838808': ['17639753', '17639981'], 'ENSE00003522686': ['17640202', '17640302'], 'ENSE00003486711': ['17648554', '17648675'], 'ENSE00001894687': ['17654636', '17654665']}, 'ENST00000495593': {'ENSE00001823966': ['17694047', '17695280'], 'ENSE00003479368': ['17696848', '17697874']}}}\n",
      "{'PRDM16': {'ENST00000511072': {'ENSE00002048533': ['3069168', '3069296'], 'ENSE00001754112': ['3186125', '3186474'], 'ENSE00003480863': ['3244087', '3244137'], 'ENSE00002034212': ['3385149', '3385286'], 'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00003701451': ['3414560', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00002081080': ['3433677', '3434342']}, 'ENST00000607632': {'ENSE00003700259': ['3069183', '3069296'], 'ENSE00003695128': ['3186125', '3186591']}, 'ENST00000378391': {'ENSE00001222906': ['3069197', '3069296'], 'ENSE00001754112': ['3186125', '3186474'], 'ENSE00003480863': ['3244087', '3244137'], 'ENSE00003695725': ['3385152', '3385286'], 'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00003701451': ['3414560', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00003700833': ['3431966', '3432140'], 'ENSE00001477303': ['3433734', '3435421']}, 'ENST00000514189': {'ENSE00001598723': ['3069211', '3069296'], 'ENSE00001754112': ['3186125', '3186474'], 'ENSE00003480863': ['3244087', '3244137'], 'ENSE00002034212': ['3385149', '3385286'], 'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00001582537': ['3414563', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00002081080': ['3433677', '3434342']}, 'ENST00000270722': {'ENSE00003850248': ['3069203', '3069296'], 'ENSE00001754112': ['3186125', '3186474'], 'ENSE00003480863': ['3244087', '3244137'], 'ENSE00003695725': ['3385152', '3385286'], 'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00003701451': ['3414560', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00003700833': ['3431966', '3432140'], 'ENSE00001759023': ['3433677', '3438621']}, 'ENST00000512462': {'ENSE00002087812': ['3237931', '3238095'], 'ENSE00003516886': ['3244087', '3244137'], 'ENSE00003662380': ['3385152', '3385286'], 'ENSE00003697266': ['3396491', '3396593'], 'ENSE00003701239': ['3402791', '3402998'], 'ENSE00003700436': ['3404739', '3404886'], 'ENSE00003695760': ['3405495', '3405648'], 'ENSE00003699089': ['3411384', '3412800'], 'ENSE00003702178': ['3414560', '3414647'], 'ENSE00003701293': ['3417828', '3417997'], 'ENSE00003697785': ['3418667', '3418744'], 'ENSE00003701636': ['3425581', '3425750'], 'ENSE00003696062': ['3426051', '3426225'], 'ENSE00003697832': ['3430872', '3431108'], 'ENSE00003698886': ['3431966', '3432140'], 'ENSE00002068021': ['3433677', '3433925']}, 'ENST00000463591': {'ENSE00001882136': ['3244132', '3244137'], 'ENSE00003695725': ['3385152', '3385286'], 'ENSE00001846205': ['3386676', '3386918']}, 'ENST00000509860': {'ENSE00003700221': ['3396491', '3396593'], 'ENSE00003696962': ['3402791', '3402998'], 'ENSE00003700688': ['3404739', '3404886'], 'ENSE00003700645': ['3405495', '3405648'], 'ENSE00003695658': ['3411384', '3412800'], 'ENSE00001582537': ['3414563', '3414647'], 'ENSE00003699052': ['3417828', '3417997'], 'ENSE00003698430': ['3418667', '3418744'], 'ENSE00003699796': ['3425581', '3425750'], 'ENSE00003701891': ['3426051', '3426225'], 'ENSE00003698226': ['3430872', '3431108'], 'ENSE00003700833': ['3431966', '3432140'], 'ENSE00001477317': ['3433677', '3434293']}, 'ENST00000378389': {'ENSE00001905893': ['3424919', '3425059'], 'ENSE00003701636': ['3425581', '3425750'], 'ENSE00003696062': ['3426051', '3426225'], 'ENSE00003697832': ['3430872', '3431108'], 'ENSE00001875784': ['3433677', '3434095']}, 'ENST00000606170': {'ENSE00003697875': ['3425216', '3425750'], 'ENSE00003700534': ['3426051', '3426072']}}, 'THRAP3': {'ENST00000354618': {'ENSE00001459618': ['36224443', '36224505'], 'ENSE00001354450': ['36259382', '36259484'], 'ENSE00001298923': ['36282533', '36282700'], 'ENSE00001291272': ['36286368', '36287270'], 'ENSE00000955947': ['36289060', '36289764'], 'ENSE00003566879': ['36291374', '36291546'], 'ENSE00003479354': ['36292598', '36292709'], 'ENSE00003660836': ['36293851', '36293935'], 'ENSE00000766231': ['36296583', '36296770'], 'ENSE00001079145': ['36300886', '36301084'], 'ENSE00001079144': ['36301553', '36301696'], 'ENSE00001402345': ['36303796', '36305357']}, 'ENST00000469141': {'ENSE00001829811': ['36224432', '36224505'], 'ENSE00001910590': ['36224971', '36225102'], 'ENSE00001354450': ['36259382', '36259484'], 'ENSE00001298923': ['36282533', '36282700'], 'ENSE00001291272': ['36286368', '36287270'], 'ENSE00000955947': ['36289060', '36289764'], 'ENSE00003566879': ['36291374', '36291546'], 'ENSE00003479354': ['36292598', '36292709'], 'ENSE00003660836': ['36293851', '36293935'], 'ENSE00000766231': ['36296583', '36296770'], 'ENSE00001079145': ['36300886', '36301084'], 'ENSE00001079144': ['36301553', '36301696'], 'ENSE00002179918': ['36303796', '36304148']}, 'ENST00000648638': {'ENSE00001861135': ['36224433', '36224505'], 'ENSE00001917968': ['36225412', '36225463'], 'ENSE00001354450': ['36259382', '36259484'], 'ENSE00001298923': ['36282533', '36282700'], 'ENSE00001291272': ['36286368', '36287270'], 'ENSE00000955947': ['36289060', '36289764'], 'ENSE00003566879': ['36291374', '36291546'], 'ENSE00003479354': ['36292598', '36292709'], 'ENSE00003660836': ['36293851', '36293935'], 'ENSE00000766231': ['36296583', '36296770'], 'ENSE00001079145': ['36300886', '36301084'], 'ENSE00001079144': ['36301553', '36301696'], 'ENSE00003840935': ['36303796', '36303831'], 'ENSE00003839481': ['36304474', '36305357']}, 'ENST00000466743': {'ENSE00001941050': ['36289718', '36289764'], 'ENSE00003588907': ['36291374', '36291546'], 'ENSE00003608584': ['36292598', '36292709'], 'ENSE00003570214': ['36293851', '36293935'], 'ENSE00001859170': ['36294050', '36294169']}}, 'CDKN2C': {'ENST00000262662': {'ENSE00001248960': ['50960745', '50960803'], 'ENSE00001248955': ['50967936', '50969899'], 'ENSE00001064516': ['50970358', '50970497'], 'ENSE00001064515': ['50973893', '50974633']}, 'ENST00000396148': {'ENSE00001524054': ['50968706', '50969899'], 'ENSE00001064516': ['50970358', '50970497'], 'ENSE00001456048': ['50973893', '50974634']}, 'ENST00000371761': {'ENSE00001456052': ['50970245', '50970497'], 'ENSE00001456048': ['50973893', '50974634']}}, 'TENT5C': {'ENST00000369448': {'ENSE00001450051': ['117606048', '117606153'], 'ENSE00001450050': ['117622842', '117628389']}}, 'TPR': {'ENST00000367478': {'ENSE00003843206': ['186374878', '186375253'], 'ENSE00003487734': ['186373359', '186373463'], 'ENSE00003587929': ['186370970', '186371043'], 'ENSE00003662167': ['186367886', '186367982'], 'ENSE00003522190': ['186363342', '186363445'], 'ENSE00003577455': ['186362837', '186363001'], 'ENSE00003490606': ['186362288', '186362380'], 'ENSE00003604606': ['186361789', '186361869'], 'ENSE00003514225': ['186361622', '186361709'], 'ENSE00003535023': ['186360765', '186360905'], 'ENSE00003689727': ['186360273', '186360364'], 'ENSE00003477153': ['186359799', '186359996'], 'ENSE00003645437': ['186358543', '186358650'], 'ENSE00003608274': ['186357397', '186357623'], 'ENSE00003685897': ['186356286', '186356449'], 'ENSE00003582311': ['186355635', '186355768'], 'ENSE00003610313': ['186355410', '186355558'], 'ENSE00003670368': ['186353688', '186353850'], 'ENSE00003582315': ['186351976', '186352110'], 'ENSE00001030868': ['186351330', '186351470'], 'ENSE00001030871': ['186350223', '186350388'], 'ENSE00001030911': ['186347292', '186347458'], 'ENSE00001030867': ['186346135', '186346287'], 'ENSE00001068225': ['186345580', '186345696'], 'ENSE00001068221': ['186344375', '186344578'], 'ENSE00003529386': ['186343906', '186344090'], 'ENSE00001265722': ['186343326', '186343473'], 'ENSE00001068222': ['186341252', '186341389'], 'ENSE00003623500': ['186341028', '186341159'], 'ENSE00003615941': ['186339642', '186339772'], 'ENSE00001030928': ['186338033', '186338243'], 'ENSE00001030892': ['186337013', '186337156'], 'ENSE00001030903': ['186336496', '186336694'], 'ENSE00001030889': ['186335338', '186335543'], 'ENSE00001030922': ['186335068', '186335129'], 'ENSE00001030902': ['186334325', '186334533'], 'ENSE00001030886': ['186333122', '186333394'], 'ENSE00001030900': ['186332195', '186332343'], 'ENSE00001030927': ['186331498', '186331581'], 'ENSE00001444649': ['186327460', '186327660'], 'ENSE00001030908': ['186326104', '186326235'], 'ENSE00000790831': ['186325764', '186325854'], 'ENSE00000790830': ['186323686', '186323870'], 'ENSE00000790829': ['186322518', '186322586'], 'ENSE00003507780': ['186322318', '186322412'], 'ENSE00003525673': ['186320312', '186320418'], 'ENSE00003691190': ['186318733', '186318828'], 'ENSE00000790825': ['186318447', '186318603'], 'ENSE00000790824': ['186317482', '186317600'], 'ENSE00000823022': ['186314629', '186314724'], 'ENSE00001444646': ['186311652', '186314026']}, 'ENST00000467810': {'ENSE00001945374': ['186322518', '186322701'], 'ENSE00003499052': ['186322318', '186322412'], 'ENSE00003629978': ['186320312', '186320418'], 'ENSE00003562529': ['186318733', '186318828'], 'ENSE00001817070': ['186318505', '186318603']}, 'ENST00000492973': {'ENSE00001871801': ['186326104', '186326220'], 'ENSE00001887741': ['186325519', '186325854']}, 'ENST00000481347': {'ENSE00003655025': ['186343906', '186344090'], 'ENSE00001945728': ['186341252', '186343473'], 'ENSE00003528301': ['186341028', '186341159'], 'ENSE00003549857': ['186339642', '186339772'], 'ENSE00001842188': ['186338187', '186338243']}, 'ENST00000491783': {'ENSE00001901888': ['186356286', '186356538'], 'ENSE00003466865': ['186355635', '186355768'], 'ENSE00003471722': ['186355410', '186355558'], 'ENSE00003505085': ['186353688', '186353850'], 'ENSE00003511424': ['186351976', '186352110'], 'ENSE00001934432': ['186351423', '186351470']}, 'ENST00000469463': {'ENSE00001948608': ['186353688', '186353948'], 'ENSE00001831515': ['186351952', '186352110']}, 'ENST00000474852': {'ENSE00003677161': ['186374878', '186375325'], 'ENSE00003490929': ['186373359', '186373463'], 'ENSE00003620212': ['186370970', '186371043'], 'ENSE00003545532': ['186367886', '186367982'], 'ENSE00003611427': ['186363342', '186363445'], 'ENSE00003651485': ['186362837', '186363001'], 'ENSE00003638079': ['186362288', '186362380'], 'ENSE00003527111': ['186361789', '186361869'], 'ENSE00003517331': ['186361622', '186361709'], 'ENSE00003594539': ['186360765', '186360905'], 'ENSE00003693781': ['186360273', '186360364'], 'ENSE00003689787': ['186359799', '186359996'], 'ENSE00003619958': ['186358543', '186358650'], 'ENSE00003478760': ['186357397', '186357623'], 'ENSE00003662843': ['186356286', '186356449'], 'ENSE00003466865': ['186355635', '186355768'], 'ENSE00003471722': ['186355410', '186355558'], 'ENSE00001893692': ['186353688', '186353880'], 'ENSE00001896671': ['186352015', '186352110']}, 'ENST00000451586': {'ENSE00001617725': ['186375559', '186375693'], 'ENSE00001068219': ['186374878', '186375186'], 'ENSE00001753818': ['186373393', '186373463']}, 'ENST00000613151': {'ENSE00003466314': ['186374878', '186375325'], 'ENSE00003487734': ['186373359', '186373463'], 'ENSE00003587929': ['186370970', '186371043'], 'ENSE00003662167': ['186367886', '186367982'], 'ENSE00003522190': ['186363342', '186363445'], 'ENSE00003577455': ['186362837', '186363001'], 'ENSE00003490606': ['186362288', '186362380'], 'ENSE00003604606': ['186361789', '186361869'], 'ENSE00003514225': ['186361622', '186361709'], 'ENSE00003535023': ['186360765', '186360905'], 'ENSE00003689727': ['186360273', '186360364'], 'ENSE00003477153': ['186359799', '186359996'], 'ENSE00003645437': ['186358543', '186358650'], 'ENSE00003608274': ['186357397', '186357623'], 'ENSE00003685897': ['186356286', '186356449'], 'ENSE00003582311': ['186355635', '186355768'], 'ENSE00003610313': ['186355410', '186355558'], 'ENSE00003733373': ['186353871', '186353880']}}, 'PRRX1': {'ENST00000553786': {'ENSE00002433506': ['170662728', '170662924'], 'ENSE00001814256': ['170664306', '170664459'], 'ENSE00002386431': ['170719726', '170719903']}, 'ENST00000367760': {'ENSE00001907015': ['170663134', '170664459'], 'ENSE00002347336': ['170719726', '170719901'], 'ENSE00003617120': ['170726220', '170726401'], 'ENSE00003697545': ['170730277', '170730348'], 'ENSE00001834707': ['170736048', '170736508']}, 'ENST00000239461': {'ENSE00001445579': ['170664131', '170664459'], 'ENSE00002347336': ['170719726', '170719901'], 'ENSE00003617120': ['170726220', '170726401'], 'ENSE00001821799': ['170736048', '170739421']}, 'ENST00000497230': {'ENSE00002447784': ['170663937', '170664459'], 'ENSE00002347336': ['170719726', '170719901'], 'ENSE00001919426': ['170726220', '170729021']}, 'ENST00000476867': {'ENSE00001873292': ['170722617', '170722668'], 'ENSE00003530720': ['170726220', '170726401'], 'ENSE00002517633': ['170736048', '170736399']}, 'ENST00000495280': {'ENSE00001936632': ['170722645', '170722668'], 'ENSE00003530720': ['170726220', '170726401'], 'ENSE00003697845': ['170730277', '170730348'], 'ENSE00001913921': ['170736048', '170736233']}, 'ENST00000485529': {'ENSE00001889459': ['170730206', '170730348'], 'ENSE00001913921': ['170736048', '170736233']}, 'ENST00000496573': {'ENSE00001941384': ['170730446', '170730621'], 'ENSE00001913921': ['170736048', '170736233']}}, 'ARHGEF10L': {'ENST00000361221': {'ENSE00001267745': ['17539698', '17539950'], 'ENSE00001162086': ['17580553', '17580632'], 'ENSE00001161931': ['17587460', '17587645'], 'ENSE00001182858': ['17588446', '17588479'], 'ENSE00003649194': ['17602127', '17602218'], 'ENSE00003548497': ['17603508', '17603591'], 'ENSE00003670661': ['17607802', '17607977'], 'ENSE00000896633': ['17613058', '17613174'], 'ENSE00000896634': ['17616094', '17616202'], 'ENSE00003503885': ['17619339', '17619445'], 'ENSE00003505774': ['17621864', '17621941'], 'ENSE00003599477': ['17622996', '17623175'], 'ENSE00003493681': ['17624387', '17624503'], 'ENSE00003641913': ['17625956', '17626048'], 'ENSE00003583343': ['17627330', '17627503'], 'ENSE00003654161': ['17632321', '17632466'], 'ENSE00001755595': ['17634548', '17634562'], 'ENSE00003461835': ['17634835', '17635016'], 'ENSE00003650063': ['17637888', '17638003'], 'ENSE00003685080': ['17638562', '17638689'], 'ENSE00003669483': ['17640202', '17640302'], 'ENSE00003631164': ['17648554', '17648675'], 'ENSE00003678534': ['17654636', '17654722'], 'ENSE00003507371': ['17655879', '17656102'], 'ENSE00003468102': ['17656554', '17656708'], 'ENSE00003607478': ['17664447', '17664595'], 'ENSE00003506428': ['17687573', '17687747'], 'ENSE00003460282': ['17695158', '17695280'], 'ENSE00001954969': ['17696848', '17697869']}, 'ENST00000375415': {'ENSE00001162086': ['17580553', '17580632'], 'ENSE00001161931': ['17587460', '17587645'], 'ENSE00001182858': ['17588446', '17588479'], 'ENSE00003649194': ['17602127', '17602218'], 'ENSE00003548497': ['17603508', '17603591'], 'ENSE00003670661': ['17607802', '17607977'], 'ENSE00000896634': ['17616094', '17616202'], 'ENSE00003503885': ['17619339', '17619445'], 'ENSE00003505774': ['17621864', '17621941'], 'ENSE00003599477': ['17622996', '17623175'], 'ENSE00003493681': ['17624387', '17624503'], 'ENSE00003641913': ['17625956', '17626048'], 'ENSE00003583343': ['17627330', '17627503'], 'ENSE00003654161': ['17632321', '17632466'], 'ENSE00001755595': ['17634548', '17634562'], 'ENSE00003461835': ['17634835', '17635016'], 'ENSE00003650063': ['17637888', '17638003'], 'ENSE00003685080': ['17638562', '17638689'], 'ENSE00003669483': ['17640202', '17640302'], 'ENSE00003631164': ['17648554', '17648675'], 'ENSE00003678534': ['17654636', '17654722'], 'ENSE00003507371': ['17655879', '17656102'], 'ENSE00003468102': ['17656554', '17656708'], 'ENSE00003607478': ['17664447', '17664595'], 'ENSE00003506428': ['17687573', '17687747'], 'ENSE00003460282': ['17695158', '17695280'], 'ENSE00001467009': ['17696848', '17697649']}, 'ENST00000482892': {'ENSE00001850961': ['17602171', '17602218'], 'ENSE00003623586': ['17603508', '17603591'], 'ENSE00001932431': ['17604476', '17604882']}, 'ENST00000469726': {'ENSE00001855212': ['17615088', '17616202'], 'ENSE00003616000': ['17619339', '17619445'], 'ENSE00003668894': ['17621864', '17621941'], 'ENSE00003594571': ['17622996', '17623175'], 'ENSE00003555762': ['17624387', '17624503'], 'ENSE00003659199': ['17625956', '17626048'], 'ENSE00003658233': ['17627330', '17627503'], 'ENSE00003686475': ['17632321', '17632466'], 'ENSE00003558496': ['17634835', '17635016'], 'ENSE00003602420': ['17637888', '17638003'], 'ENSE00003576297': ['17638562', '17638689'], 'ENSE00003522686': ['17640202', '17640302'], 'ENSE00003486711': ['17648554', '17648675'], 'ENSE00003515029': ['17654636', '17654722'], 'ENSE00003685692': ['17655879', '17656102'], 'ENSE00003466542': ['17656554', '17656708'], 'ENSE00003468557': ['17664447', '17664595'], 'ENSE00003526832': ['17687573', '17687747'], 'ENSE00003499717': ['17695158', '17695280'], 'ENSE00001889352': ['17696848', '17697860']}, 'ENST00000375408': {'ENSE00001935881': ['17618316', '17618522'], 'ENSE00003503885': ['17619339', '17619445'], 'ENSE00003505774': ['17621864', '17621941'], 'ENSE00003599477': ['17622996', '17623175'], 'ENSE00003493681': ['17624387', '17624503'], 'ENSE00003641913': ['17625956', '17626048'], 'ENSE00003583343': ['17627330', '17627503'], 'ENSE00003654161': ['17632321', '17632466'], 'ENSE00003461835': ['17634835', '17635016'], 'ENSE00003650063': ['17637888', '17638003'], 'ENSE00003685080': ['17638562', '17638689'], 'ENSE00003669483': ['17640202', '17640302'], 'ENSE00003631164': ['17648554', '17648675'], 'ENSE00003678534': ['17654636', '17654722'], 'ENSE00003507371': ['17655879', '17656102'], 'ENSE00003468102': ['17656554', '17656708'], 'ENSE00003607478': ['17664447', '17664595'], 'ENSE00003506428': ['17687573', '17687747'], 'ENSE00003460282': ['17695158', '17695280'], 'ENSE00003544025': ['17696848', '17697868']}, 'ENST00000167825': {'ENSE00001833298': ['17626007', '17626048'], 'ENSE00003583343': ['17627330', '17627503'], 'ENSE00003654161': ['17632321', '17632466'], 'ENSE00003461835': ['17634835', '17635016'], 'ENSE00003650063': ['17637888', '17638003'], 'ENSE00003685080': ['17638562', '17638689'], 'ENSE00003669483': ['17640202', '17640302'], 'ENSE00003631164': ['17648554', '17648675'], 'ENSE00003678534': ['17654636', '17654722'], 'ENSE00003507371': ['17655879', '17656102'], 'ENSE00003468102': ['17656554', '17656708'], 'ENSE00003607478': ['17664447', '17664595'], 'ENSE00003506428': ['17687573', '17687747'], 'ENSE00001952719': ['17691120', '17691192'], 'ENSE00003499717': ['17695158', '17695280'], 'ENSE00003546919': ['17696848', '17697868']}, 'ENST00000482359': {'ENSE00001820929': ['17634257', '17634562'], 'ENSE00003558496': ['17634835', '17635016'], 'ENSE00003602420': ['17637888', '17638003']}, 'ENST00000466782': {'ENSE00001838808': ['17639753', '17639981'], 'ENSE00003522686': ['17640202', '17640302'], 'ENSE00003486711': ['17648554', '17648675'], 'ENSE00001894687': ['17654636', '17654665']}, 'ENST00000495593': {'ENSE00001823966': ['17694047', '17695280'], 'ENSE00003479368': ['17696848', '17697874']}}}\n"
     ]
    }
   ],
   "source": [
    "# FAIRE LE DICTIONNAIRE GENE_DIC/GENE_DIC2\n",
    "\n",
    "# Ce dictionnaire contient tous les gènes présents dans le catalogue cosmics. Tous les transcrits du gènes présents dans le GTF et tous les exons\n",
    "# faisant partie de chaque transcrit, ainsi que leur position de départ et de fin sont ajoutés dans le dictionnaire.\n",
    "\n",
    "#count = 0 # Sert à arrêter le code (Ne pas générer les 706 gènes)\n",
    "import re\n",
    "nature_du_brin = {} \n",
    "gene_dic = {}\n",
    "data = '../data/Homo_sapiens.GRCh38.113.gtf' # GTF\n",
    "with open(data, 'r') as file:\n",
    "    for line in file:\n",
    "        # Séparation dans les lignes à l'aide des caractères '\\t' et '\"'\n",
    "        colonne_tab = line.split('\\t') \n",
    "        colonne_guillemet = line.split('\"')\n",
    "            \n",
    "        if len(colonne_tab) >= 3 and len(colonne_guillemet) >= 3: # Retire les trois premières lignes du file qui n'ont pas rapport\n",
    "\n",
    "            if colonne_tab[2]== 'gene': # Seulement les lignes des gènes\n",
    "                gene = colonne_guillemet[1] # Identifiant du gène ENSG\n",
    "                \n",
    "                for valeur in equivalence_nom.values(): # Itération sur les identifiants ENSG qui font parties du catalogue COSMIC\n",
    "                    if valeur == gene:\n",
    "                        count += 1\n",
    "                        brand = colonne_tab[6] # Nature du brin du gène\n",
    "                        nature_du_brin[gene]= brand # Dictionnaire qui relie le nom du gène ENSG à la nature du brin sur lequel il se retrouve\n",
    "                        gene_dic[gene]= {} # Ajoute le nom du gène ENSG comme clé dans le gene_dic\n",
    "            \n",
    "            # Itération sur chaque gène ajouté dans le gene_dic (ça veut dire qu'il fait partie du catalogue)     \n",
    "            for gene in gene_dic:\n",
    "                if gene in line: \n",
    "                    if colonne_tab[2]== 'transcript': \n",
    "                        transcrit_nom = colonne_guillemet[5] # Nom du transcrit ENST\n",
    "                        gene = colonne_guillemet[1] # Nom du gène ENSG\n",
    "                        gene_dic[gene][transcrit_nom]= {} # Ajoute les transcrits sous les bons gènes\n",
    "                    if colonne_tab[2]== 'exon':\n",
    "                        gene = colonne_guillemet[1] # Nom du gène ENSG\n",
    "                        transcrit_name_info = re.search(r'ENST[0123456789]{5,100}', line)\n",
    "                        transcrit_name = transcrit_name_info.group(0) # Nom du transcrit ENST\n",
    "                        exon_name_info = re.search(r'ENSE[0123456789]{5,100}', line)\n",
    "                        exon_name = exon_name_info.group(0) # Nom de l'exon ENSE\n",
    "                \n",
    "                        # Positions de départ et de fin de l'exon\n",
    "                        petit = colonne_tab[3] # Position génomique la plus petite\n",
    "                        grand = colonne_tab[4] # Position génomique la plus grande\n",
    "\n",
    "                        for key in gene_dic[gene].keys(): # Itère sur les transcrits faisant partie d'un gène\n",
    "                            if key == transcrit_name: \n",
    "                                gene_dic[gene][transcrit_name][exon_name]= [] # Ajoute les exons dans les bons transcrits\n",
    "\n",
    "                        for key in gene_dic[gene][transcrit_name].keys(): # Itère sur les exons faisant partie des transcrits\n",
    "                            if key == exon_name: \n",
    "                                \n",
    "                                # Ajoutes les positions de début et de fin de l'exon\n",
    "                                gene_dic[gene][transcrit_name][exon_name].append(petit) \n",
    "                                gene_dic[gene][transcrit_name][exon_name].append(grand)\n",
    "\n",
    "                                \n",
    "# Il serait pratique d'avoir un gene_dic2 qui utilise le nom usuel du gène (ex. BRCA1) au lieu de son identifiant ENSG, car la base COSMICS n'utilise\n",
    "# pas l'identifiant ENSG.\n",
    "\n",
    "modif_equivalence_nom = {valeur: cle for cle, valeur in equivalence_nom.items()} # Cette ligne permet d'inverser les clés et les valeurs dans le dictionnaire equivalence_nom\n",
    "nature_du_brin2 = {modif_equivalence_nom.get(cle, cle): valeur for cle, valeur in nature_du_brin.items()}\n",
    "gene_dic2 = {modif_equivalence_nom.get(cle, cle): valeur for cle, valeur in gene_dic.items()}\n",
    "\n",
    "\n",
    "##### À RETIRER LORSQUE LE SCRIPT SERA TERMINÉ\n",
    "#gene_dic.popitem()\n",
    "#gene_dic2.popitem()\n",
    "\n",
    "\n",
    "#print(nature_du_brin)\n",
    "#print(nature_du_brin2)\n",
    "#print(gene_dic) # Avec identifiant ENSG des gènes\n",
    "#print(gene_dic2) # Avec nom usuel COSMICS des gènes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e42cba5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Erreur! Le nombre de gène à traiter est de 706 et le gene_dic contient 7 gènes\n"
     ]
    }
   ],
   "source": [
    "# VÉRIFICATION DU GENE_DIC ET GENE_DIC2 GÉNÉRÉ\n",
    "# Permet de vérifier que le nombre de gène généré dans le gene_dic est égale à celui dans la database de Cosmics\n",
    "\n",
    "if len(equivalence_nom)== len(gene_dic2):\n",
    "    print('Correct! Le nombre de gène à traiter est de ' + str(len(equivalence_nom)) + ' et le gene_dic contient ' + str(len(gene_dic)) + ' gènes')\n",
    "else:\n",
    "    print('Erreur! Le nombre de gène à traiter est de ' + str(len(equivalence_nom)) + ' et le gene_dic contient ' + str(len(gene_dic)) + ' gènes')\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "41417ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sequence_cDNA (gene, transcrit):\n",
    "    \"\"\"\n",
    "    Sortir la séquence d'ADNc d'un transcrit d'un gène.\n",
    "    \n",
    "    Paramètres\n",
    "    ----------\n",
    "        gene : nom du gène (ENSG)\n",
    "        transcrit : nom du transcrit (ENST)\n",
    "            \n",
    "    Returns\n",
    "    -------\n",
    "        la séquence d'ADNc du transcrit d'un gène\n",
    "    \"\"\"\n",
    "    header_of_transcrit = ''\n",
    "\n",
    "    with open('../data/Homo_sapiens.GRCh38.cdna.all.fa', 'r') as file:  # Toutes les séquences ADNc de chaque transcrit de tous les gènes\n",
    "        for line in file:\n",
    "            if gene in line and transcrit in line:\n",
    "                header_of_transcrit = line.strip() # Stock le header de la séquence recherché\n",
    "                break \n",
    "\n",
    "    if header_of_transcrit:\n",
    "        sequence = \"\" \n",
    "        header_trouve = False\n",
    "\n",
    "        with open('../data/Homo_sapiens.GRCh38.cdna.all.fa', 'r') as file:\n",
    "            for line in file:\n",
    "                line = line.strip()\n",
    "                if line.startswith(\">\"):\n",
    "                    if header_trouve:\n",
    "                        break\n",
    "\n",
    "                    if line == header_of_transcrit:\n",
    "                        header_trouve = True\n",
    "\n",
    "                elif header_trouve:\n",
    "                    sequence += line # Stock la séquence d'ADNc du transcrit\n",
    "\n",
    "        if sequence: # Si la séquence a été trouvé\n",
    "            return sequence\n",
    "            \n",
    "        else: # Si aucune séquence n'a été trouvé\n",
    "            return (\"Séquence non trouvée.\")\n",
    "            \n",
    "    else: # Si aucun header n'a été trouvé\n",
    "        return (f\"Le header {header_of_transcrit} n'a pas été trouvé.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d8321ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_position_chronologique(gene, position_chromosomique, transcrit): \n",
    "    \"\"\"\n",
    "    Convertir une position génomique en position transcriptomique. N.B. première position = 1.\n",
    "    \n",
    "    Paramètres\n",
    "    ----------\n",
    "        gene : nom du gène (ENSG)\n",
    "        position_chromosomique : position dans le génome\n",
    "        transcrit : nom du transcrit (ENST)\n",
    "            \n",
    "    Returns\n",
    "    -------\n",
    "        la position transcriptomique\n",
    "    \"\"\"    \n",
    "    total = 0\n",
    "    liste_position_fin_exon = [0] # Liste qui va contenir la position de fin de chaque exon en coordonnées chronologiques\n",
    "\n",
    "    for exon in gene_dic[gene][transcrit]:\n",
    "        \n",
    "        # Calcul de la longueur de l'exon en nombre de nucléotides\n",
    "        brand = nature_du_brin[gene]\n",
    "        longueur_exon = int(gene_dic[gene][transcrit][exon][1])-int(gene_dic[gene][transcrit][exon][0])+1 \n",
    "        total = total + int(longueur_exon) # Cette variable finit par contenir la longueur de l'exon et tous les exons précédents = position chronologique de fin pour chaque exon\n",
    "        liste_position_fin_exon.append(total) # Ajoute les positions chronologiques de fin pour chaque exon\n",
    "\n",
    "        # CONDITIONS: la position génomique doit être dans un exon\n",
    "        if position_chromosomique >= int(gene_dic[gene][transcrit][exon][0]) and position_chromosomique <= int(gene_dic[gene][transcrit][exon][1]): \n",
    "            keys = list(gene_dic[gene][transcrit].keys()) # Stock le nom de l'exon ENSE présentemment itéré\n",
    "            position_keys = keys.index(exon) # Stock position de l'exon dans le transcrit\n",
    "\n",
    "            # Pour un gène sur le brin négatif\n",
    "            if brand == '-':\n",
    "                # Calcul de la conversion des coordonnées chromosomiques en coordonnées chronologiques dans un transcrit\n",
    "                position_chronologique = int(liste_position_fin_exon[position_keys])+1+(int(gene_dic[gene][transcrit][exon][1])-position_chromosomique) \n",
    "\n",
    "            # Pour un gène sur le brin positif\n",
    "            if brand == '+':\n",
    "                # Calcul de la conversion des coordonnées chromosomiques en coordonnées chronologiques dans un transcrit\n",
    "                position_chronologique = int(liste_position_fin_exon[position_keys])+1+(position_chromosomique-int(gene_dic[gene][transcrit][exon][0]))\n",
    "    return position_chronologique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b4da0765",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RÉPARTITION DES MUTATIONS SI ELLES TOMBENT DANS UN EXON, DANS UN CHEVAUCHEMENT INTRON-EXON OU DANS UN INTRON ET DE QUEL TYPE DE MUTATION IL S'AGIT ( SUBS, INV, INS, DEL, DELINS, DUP)\n",
    "\n",
    "# Triage des mutations pour chaque gène\n",
    "\n",
    "# Cette première section sert à supprimer le contenu existant déjà dans le fichier si nécessaire \n",
    "mutation = ['del_ins', 'deletion', 'duplication', 'insertion', 'substitution', 'chevauchement', 'rejected', 'autre', 'unknow', 'inversion', 'incomplete'] # Cette liste contient tous les types de mutations\n",
    "for element in mutation: # Itération sur chaque élément présent dans la liste\n",
    "    nom_fichier = f\"../result/mutation_{element}.tsv\"\n",
    "    with open(nom_fichier, mode='w') as file:\n",
    "        file.write('') # Supprime le contenu du fichier\n",
    "\n",
    "\n",
    "ligne_ajoutee = set() # Contrôle que chaque mutation est traité une seule fois, malgré toutes les loops\n",
    "data = '../data/Cosmic_MutantCensus_v101_GRCh38.tsv' # Catalogue COSMICS: toutes les mutations de tous les gènes\n",
    "with open(data, 'r') as file: \n",
    "    for line in file: \n",
    "        colonne = line.split('\\t') \n",
    "        for gene_name in gene_dic2: # Itération sur les noms de gène COSMICS\n",
    "            if colonne[0] == gene_name: # Conserve les lignes sur le gène présentement itéré\n",
    "                mutation_type = colonne[11] # missense, nonsense, inframe, frameshift, etc.\n",
    "                transcrit_reference = les_transcrits_de_reference[gene_name]\n",
    "                mutation_grand = colonne[16] # Grande position CHROMOSOMIQUE de la mutation\n",
    "                mutation_petit = colonne[15] # Petite position CHROMOSOMIQUE de la mutation\n",
    "                mutation_explain = colonne[9] # Description de la mutation \n",
    "                mutation_id = colonne[8] # Identifiant de la mutation (numéros)\n",
    "                mutation_descrip = colonne[22] # Autre descriptif de la mutation, utile pour les duplications\n",
    "                \n",
    "                # Gérer les mutations Unknow: elles sont retirés puisqu'elles ne possèdent pas assez d'information pour être exécutée\n",
    "                if mutation_grand == '' or mutation_petit == '':\n",
    "                    nom_fichier = \"../result/mutation_unknow.tsv\"\n",
    "                    with open(nom_fichier, mode='a') as unknow:\n",
    "                        unknow.write(line)\n",
    "                        \n",
    "                else:\n",
    "                    \n",
    "                    # Séparation des mutations qui tombent dans un transcrit, dans les introns ou qui chevauchent les exons et les introns (souvent des splice_donor ou splice_acceptor)\n",
    "                    for exon in gene_dic2[gene_name][transcrit_reference]: # Itération sur chaque exon faisant partie du transcrit de référence\n",
    "\n",
    "                        # La mutation doit tomber dans un exon du transcrit de référence! N.B: J'enlève les mutations splice_acceptor, splice_donnor...\n",
    "                        if ((int(mutation_grand) >= int(gene_dic2[gene_name][transcrit_reference][exon][0]) and int(mutation_grand) <= int(gene_dic2[gene_name][transcrit_reference][exon][1])) and (int(mutation_petit) >= int(gene_dic2[gene_name][transcrit_reference][exon][0]) and int(mutation_petit) <= int(gene_dic2[gene_name][transcrit_reference][exon][1]))) and not mutation_id in ligne_ajoutee and 'splice' not in mutation_type:\n",
    "\n",
    "                            # Séparation des types de mutations. Les types de mutations sont triés dans des fichiers .tsv commun pour tous les gènes\n",
    "                            # Si la mutation est une substitution:\n",
    "                            if '>' in mutation_explain:\n",
    "                                nom_fichier = \"../result/mutation_substitution.tsv\"\n",
    "                                with open(nom_fichier, mode='a') as f:\n",
    "                                    f.write(line)\n",
    "                                    ligne_ajoutee.add(mutation_id) # Marquage\n",
    "                                \n",
    "                                    \n",
    "                            # Si la mutation est une délétion:    \n",
    "                            elif mutation_explain.endswith('del'):\n",
    "                                nom_fichier = f\"../result/mutation_deletion.tsv\"\n",
    "                                with open(nom_fichier, mode='a') as fichier:\n",
    "                                    fichier.write(line)\n",
    "                                    ligne_ajoutee.add(mutation_id) # Marquage\n",
    "                                \n",
    "                            \n",
    "                            # Si la mutation est une duplication:\n",
    "                            elif mutation_explain.endswith('dup'):\n",
    "\n",
    "                                # Duplication de plusieurs nucléotides\n",
    "                                if '_' in mutation_descrip:\n",
    "                                    match = re.search(r'g\\.(\\d+)_([\\d]+)', mutation_explain) # Recherche la position chromosomique de départ et de fin des nucléotides dupliqués\n",
    "                                    if match:\n",
    "                                        small_pos = int(match.group(1)) # Petite position chromosomique \n",
    "                                        big_pos = int(match.group(2)) # Grande position chromosomique\n",
    "                                        for exon in gene_dic2[gene_name][transcrit_reference]:\n",
    "                                            if (small_pos >= int(gene_dic2[gene_name][transcrit_reference][exon][0]) and small_pos <= int(gene_dic2[gene_name][transcrit_reference][exon][1])) and (big_pos >= int(gene_dic2[gene_name][transcrit_reference][exon][0]) and big_pos <= int(gene_dic2[gene_name][transcrit_reference][exon][1])) and not mutation_id in ligne_ajoutee:\n",
    "                                                nom_fichier = \"../result/mutation_duplication.tsv\"\n",
    "                                                with open(nom_fichier, mode='a') as document:\n",
    "                                                    document.write(line)\n",
    "                                                    ligne_ajoutee.add(mutation_id) # Marquage\n",
    "                                            else:\n",
    "                                                if not mutation_id in ligne_ajoutee:\n",
    "                                                    nom_fichier = f\"../result/mutation_chevauchement.tsv\"\n",
    "                                                    with open(nom_fichier, mode='a') as document:\n",
    "                                                        document.write(line)\n",
    "                                                        ligne_ajoutee.add(mutation_id) # Marquage\n",
    "                            # Duplication d'un nucléotide\n",
    "                            elif not '_' in mutation_descrip:\n",
    "                                nom_fichier = f\"../result/mutation_duplication.tsv\"\n",
    "                                with open(nom_fichier, mode='a') as document:\n",
    "                                    document.write(line)\n",
    "                                    ligne_ajoutee.add(mutation_id)\n",
    "                                        \n",
    "                            # Si la mutation est une insertion:\n",
    "                            elif 'ins' in mutation_explain and not 'delins' in mutation_explain:\n",
    "                                nom_fichier = \"../result/mutation_insertion.tsv\"\n",
    "                                with open(nom_fichier, mode='a') as papier:\n",
    "                                    papier.write(line)\n",
    "                                    ligne_ajoutee.add(mutation_id) # Marquage\n",
    "\n",
    "                            # Si la mutation est une délétion-insertion:\n",
    "                            elif 'delins' in mutation_explain:\n",
    "                                nom_fichier = \"../result/mutation_del_ins.tsv\"\n",
    "                                with open(nom_fichier, mode='a') as tableau:\n",
    "                                    tableau.write(line)\n",
    "                                    ligne_ajoutee.add(mutation_id) # Marquage\n",
    "\n",
    "                            # Si la mutation est une inversion:\n",
    "                            elif mutation_explain.endswith('inv'):\n",
    "                                nom_fichier = \"../result/mutation_inversion.tsv\"\n",
    "                                with open(nom_fichier, mode='a') as fich:\n",
    "                                    fich.write(line)\n",
    "                                    ligne_ajoutee.add(mutation_id) # Marquage\n",
    "                    \n",
    "                            # Si la mutation n'entre pas dans une des catégories précédentes. En principe, si toutes les catégories ont été définies, ce fichier devrait être vide ou inexistant\n",
    "                            else:\n",
    "                                if not mutation_id in ligne_ajoutee:\n",
    "                                    nom_fichier = \"../result/mutation_autre.tsv\"\n",
    "                                    with open(nom_fichier, mode='a') as paper:\n",
    "                                        paper.write(line)\n",
    "                                        ligne_ajoutee.add(mutation_id) # Marquage\n",
    "                        \n",
    "                        else:\n",
    "                            if ((int(mutation_grand) >= int(gene_dic2[gene_name][transcrit_reference][exon][0]) and int(mutation_grand) <= int(gene_dic2[gene_name][transcrit_reference][exon][1])) or (int(mutation_petit) >= int(gene_dic2[gene_name][transcrit_reference][exon][0]) and int(mutation_petit) <= int(gene_dic2[gene_name][transcrit_reference][exon][1]))) and not mutation_id in ligne_ajoutee:\n",
    "                                nom_fichier = \"../result/mutation_chevauchement.tsv\"\n",
    "                                with open(nom_fichier, mode='a') as f:\n",
    "                                    f.write(line)\n",
    "                                    ligne_ajoutee.add(mutation_id) # Marquage \n",
    "                                \n",
    "                    else:\n",
    "                        if not mutation_id in ligne_ajoutee: # Retient les lignes qui n'ont pas été ajouté seulement\n",
    "                            nom_fichier = \"../result/mutation_rejected.tsv\"\n",
    "                            with open(nom_fichier, mode='a') as fichier:\n",
    "                                fichier.write(line)\n",
    "                                ligne_ajoutee.add(mutation_id) \n",
    "                                \n",
    "                                \n",
    "                if mutation_id not in ligne_ajoutee:\n",
    "                    nom_fichier = f'../result/mutation_incomplete.tsv'\n",
    "                    with open(nom_fichier, mode='a') as fichier:\n",
    "                        fichier.write(line)\n",
    "                        ligne_ajoutee.add(mutation_id)\n",
    "                        \n",
    "                        \n",
    "# Description du contenu des fichiers obtenus\n",
    "                      \n",
    "# substitution: mutation de type substitution\n",
    "# del_ins: mutation de type del_ins\n",
    "# deletion: mutation de type deletion\n",
    "# insertion: mutation de type insertion\n",
    "# inversion: mutation de type inversion\n",
    "# duplication: mutation de type duplication\n",
    "# autre: autres de cas de mutation non-traité, doit être vide\n",
    "# chevauchement: mutation en chevauchement intron-exon (très souvent splice_donnor, splice_acceptor)\n",
    "# unknow et incomple: pas assez de détail sur la mutation pour être effectué (ex. pas de position)\n",
    "# rejected: mutation dans les introns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d798326c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RÉPARTITION DES MUTATIONS POUVANT ÊTRE APPLIQUÉ POUR CHAQUE TRANSCRIT DES GÈNES\n",
    "\n",
    "count_les_transcrit_mute_par_gene = {}\n",
    "traited = ['duplication', 'substitution', 'insertion', 'deletion', 'inversion', 'del_ins']                 \n",
    "for gene_name in gene_dic2: \n",
    "    gene = equivalence_nom[gene_name] # Sort l'équivalent ENSG\n",
    "    count_les_transcrit_mute_par_gene[gene_name] = {} # Ajoute le nom usuel COSMICS comme clé\n",
    "    for transcrit in gene_dic2[gene_name]:\n",
    "        count_les_transcrit_mute_par_gene[gene_name][transcrit] = [] # Ajoute les transcrits sous chaque gène\n",
    "    \n",
    "        for item in traited:\n",
    "            with open(f'../result/mutation_{item}.tsv', 'r') as file : # Ouverture des fichiers contenant les mutations triés     \n",
    "                for line in file:\n",
    "                    colonne = line.split('\\t')\n",
    "                    if colonne[0] == gene_name: # Conserve les lignes sur le gène présentement itéré\n",
    "                        mutation_petite = int(colonne[15]) # Petite position chromosomique de la mutation\n",
    "                        mutation_grande = int(colonne[16]) # Grande position chromosomique de la mutation\n",
    "                        mutation_id = colonne[8] # Identifiant de la mutation (numéro)\n",
    "                        \n",
    "                    \n",
    "                        for exon in gene_dic2[gene_name][transcrit]:\n",
    "                            # La mutation doit se retrouver dans un exon du transcrit    \n",
    "                            if (mutation_petite >= int(gene_dic2[gene_name][transcrit][exon][0]) and mutation_petite <= int(gene_dic2[gene_name][transcrit][exon][1])) and (mutation_grande >= int(gene_dic2[gene_name][transcrit][exon][0]) and mutation_grande <= int(gene_dic2[gene_name][transcrit][exon][1])):\n",
    "                                count_les_transcrit_mute_par_gene[gene_name][transcrit].append(line) # Ajoute la ligne de la mutation    \n",
    "#print(count_les_transcrit_mute_par_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "84d54edc",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### PRODUIRE CHAQUE TRANSCRIT AVEC CHAQUE MUTATION APPLICABLE #####\n",
    "# Cette section produit toutes les séquences d'ADNc des transcrits avec chaque mutation de n'importe quelle catégorie qui sont applicables.\n",
    "# Les séquences mutés sont ajoutés dans des fichiers fasta triés en fonction du type de mutation effectué \n",
    "\n",
    "\n",
    "# Cette partie sert à effacer le contenu des fichiers fasta déjà existant\n",
    "for item in traited:\n",
    "    nom_fichier = f\"../result/{item}.fasta\"\n",
    "    with open(nom_fichier, mode='w') as file: # Ouverture du fichier en mode write()\n",
    "        file.write('') # Supprime le contenu du fichier\n",
    "nom_fichier = \"../result/mutation_not_working.tsv\"\n",
    "with open(nom_fichier, mode='w') as file: # Ouverture du fichier en mode write()\n",
    "    file.write('') # Supprime le contenu du fichier\n",
    "nom_fichier = \"../result/transcrit_no_mutation.fasta\"\n",
    "with open(nom_fichier, mode='w') as file: # Ouverture du fichier en mode write()\n",
    "    file.write('') # Supprime le contenu du fichier\n",
    "nom_fichier = '../result/all_transcripts_with_mutation.fasta'\n",
    "with open(nom_fichier, mode='w') as file: # Ouverture du fichier en mode write()\n",
    "    file.write('') # Supprime le contenu du fichier\n",
    "    \n",
    "for gene_name in gene_dic2:\n",
    "    gene = equivalence_nom[gene_name] # Sort l'équivalent en ENSG\n",
    "    for transcrit in gene_dic2[gene_name]:\n",
    "        sequence = sequence_cDNA(gene, transcrit) # Sort la séquence d'ADNc\n",
    "        header1 = '>' + f'{gene}_{transcrit}' # Formation du header\n",
    "        fasta = f'../result/transcrit_no_mutation.fasta'\n",
    "        with open(fasta, 'a') as f:\n",
    "            f.write(header1 + '\\n') # Écriture du header\n",
    "            f.write(sequence + '\\n') # Écriture de la séquence\n",
    "\n",
    "lignes_deja_ajoutees = set() # Conserve les lignes de mutations défectueuses\n",
    "for gene_name in gene_dic2:\n",
    "    gene = equivalence_nom[gene_name] # Sort l'équivalent en ENSG\n",
    "    for transcrit in gene_dic2[gene_name]:\n",
    "        # TRANSCRITS SANS MUTATION\n",
    "        #if f'{gene}_{transcrit}' not in sans_mutation_fait:\n",
    "        sequence = sequence_cDNA(gene, transcrit) # Sort la séquence d'ADNc\n",
    "            #header1 = '>' + f'{gene}_{transcrit}' # Formation du header\n",
    "            #fasta = f'../result/transcrit_no_mutation.fasta'\n",
    "            #with open(fasta, 'a') as f:\n",
    "                #f.write(header1 + '\\n') # Écriture du header\n",
    "                #f.write(sequence + '\\n') # Écriture de la séquence\n",
    "                #sans_mutation_fait.add(line)\n",
    "        for line in count_les_transcrit_mute_par_gene[gene_name][transcrit]: # Itére sur chaque ligne de mutation applicable pour le transcrit\n",
    "            colonne = line.split('\\t')\n",
    "            changement = colonne[9] # Description de la mutation\n",
    "            position_petite = int(colonne[15]) # Petite position chromosomique de la mutation\n",
    "            position_grande = int(colonne[16]) # Grande position chromosomique de la mutation\n",
    "            mutation_explain = colonne[22] # Utile pour les duplications\n",
    "            mutation_id = colonne[8] # Identifiant de la mutation\n",
    "\n",
    "            # Conversion des positions chromosomiques de la mutation en position chronologique dans le transcrit COMMENÇANT PAR 0\n",
    "            position_chrono_1 = (get_position_chronologique(gene, position_petite, transcrit))-1 \n",
    "            position_chrono_2 = (get_position_chronologique(gene, position_grande, transcrit))-1\n",
    "\n",
    "            liste = [position_chrono_1, position_chrono_2] # Ajoute les deux positions chronologiques dans la liste\n",
    "            brin = colonne[17] # Nature du brin +/-\n",
    "\n",
    "            if brin == '+':\n",
    "                delete = colonne[23] # Stock la séquence nucléotidique qui est retirée\n",
    "                add = colonne[24] # Stock la séquence nucléotidique qui est ajoutée\n",
    "\n",
    "            if brin == '-': # Si le gène est sur le brin -, la database donne la séquence du brin inverse et complémentaire (brin +)\n",
    "                enleve = colonne[23]\n",
    "                chaine_inverse = enleve[::-1] # Inversion (miroir) de la chaine\n",
    "                # Obtention de la séquence complémentaire \n",
    "                adn1 = chaine_inverse.replace('A', 't') \n",
    "                adn2 = adn1.replace('T', 'a')\n",
    "                adn3 = adn2.replace('C', 'g')\n",
    "                adn4 = adn3.replace('G', 'c')\n",
    "                delete = adn4.upper()\n",
    "\n",
    "                ajoute = colonne[24]\n",
    "                chaine_miroir = ajoute[::-1] # Inversion (miroir) de la chaine\n",
    "                # Obtention de la séquence complémentaire\n",
    "                adn5 = chaine_miroir.replace('A', 't')\n",
    "                adn6 = adn5.replace('T', 'a')\n",
    "                adn7 = adn6.replace('G', 'c')\n",
    "                adn8 = adn7.replace('C', 'g')\n",
    "                add = adn8.upper()\n",
    "                \n",
    "            # GÉRER LES SUBSTITUTIONS    \n",
    "            if position_petite == position_grande and '>' in changement:\n",
    "                position_chrono = (get_position_chronologique(gene, position_petite, transcrit))-1\n",
    "                base_avant = changement[-3] # Stockage de la base initialement contenu dans le transcrit\n",
    "                base_apres = changement[-1] # Stockage de la base qui va venir remplacée l'ancienne base\n",
    "                if sequence[position_chrono]== base_avant: # Vérification\n",
    "                    sequence_modifiee = sequence[:position_chrono]+ base_apres + sequence[position_chrono+1:] # Formation de la séquence mutée\n",
    "                    header = '>' + f'{gene}_{transcrit}_{mutation_id}' # Formation du header\n",
    "                    fasta = f'../result/substitution.fasta'\n",
    "                    with open(fasta, 'a') as f:\n",
    "                        f.write(header + '\\n') # Écriture du header\n",
    "                        f.write(sequence_modifiee + '\\n') # Écriture de la séquence\n",
    "                # En théorie, si aucun problème n'est rencontré, la section suivante ne devrait pas s'effectuer\n",
    "                else: \n",
    "                    if line not in lignes_deja_ajoutees:\n",
    "                        reste = f'../result/mutation_not_working.tsv'\n",
    "                        with open(data, 'a') as r: # Ouverture du fichier qui va contenir les mutations défectueuses en mode append()\n",
    "                            r.write(line) # Écriture de la ligne dans le fichier\n",
    "                            lignes_deja_ajoutees.add(line)\n",
    "\n",
    "            \n",
    "            # GÉRER LES DELETIONS\n",
    "            elif changement.endswith('del'):\n",
    "                smallest = min(liste) # Petite position chronologique = DÉBUT DE LA MUTATION\n",
    "                biggest = max(liste) # Grande position chronologique = FIN DE LA MUTATION\n",
    "                remove_sequence = sequence[smallest:biggest+1] \n",
    "                if remove_sequence == delete: # Vérification\n",
    "                    sequence_modifiee = sequence[:smallest] + sequence[biggest+1:] # Formation de la séquence modifiée\n",
    "                    header = '>' + f'{gene}_{transcrit}_{mutation_id}' # Formation du header\n",
    "                    fasta = f'../result/deletion.fasta' \n",
    "                    with open (fasta, 'a') as f:\n",
    "                        f.write(header + '\\n') # Écriture du header\n",
    "                        f.write(sequence_modifiee + '\\n') # Écriture de la séquence\n",
    "                # En théorie, si aucun problème n'est rencontré, la section suivante ne devrait pas s'effectuer\n",
    "                else: \n",
    "                    if line not in lignes_deja_ajoutees:\n",
    "                        reste = f'../result/mutation_not_working.tsv'\n",
    "                        with open(reste, 'a') as r: # Ouverture du fichier qui récupère les mutations défectueuses en mode append()\n",
    "                            r.write(line) # Écriture de la ligne dans le fichier\n",
    "                            lignes_deja_ajoutees.add(line) \n",
    "\n",
    "            # GÉRER LES INSERTIONS\n",
    "            elif 'ins' in changement and not 'delins' in changement: \n",
    "                smallest = min(liste) # Petite position chronologique = position de départ de la mutation\n",
    "                biggest = max(liste) # Grande position chronologique = position de fin de la mutation\n",
    "                pattern = r'[ATCG]+' \n",
    "                match = re.search(pattern, changement)\n",
    "                sequence_ajoutee = match.group()\n",
    "                if sequence_ajoutee == add: # Vérification\n",
    "                    sequence_modifiee = sequence[:biggest] + sequence_ajoutee + sequence[biggest:] # Formation de la séquence mutés\n",
    "                    header = '>' + f'{gene}_{transcrit}_{mutation_id}' # Formation du header\n",
    "                    fasta = f'../result/insertion.fasta'\n",
    "                    with open (fasta, 'a') as f: \n",
    "                        f.write(header + '\\n') # Écriture du header\n",
    "                        f.write(sequence_modifiee + '\\n') # Écriture de la séquence\n",
    "                # En théorie, si aucun problème n'est rencontré, la section suivante ne devrait pas s'effectuer\n",
    "                else: \n",
    "                    if line not in lignes_deja_ajoutees: \n",
    "                        reste = f'../result/mutation_not_working.tsv'\n",
    "                        with open(reste, 'a') as r: # Ouverture du fichier qui va contenir les mutations défectueuses\n",
    "                            r.write(line) # Écriture de la ligne de la mutation\n",
    "                            lignes_deja_ajoutees.add(line)\n",
    "\n",
    "            # GÉRER LES DUPLICATIONS\n",
    "            elif changement.endswith('dup'):\n",
    "                    \n",
    "            # Le script est différent si on a une duplication d'un nucléotide ou de plusieurs nucléotides\n",
    "                # PLUSIEURS NUCLÉOTIDES DUPLIQUÉS\n",
    "                if '_' in mutation_explain:\n",
    "                    match = re.search(r'g\\.(\\d+)_([\\d]+)', mutation_explain) # Recherche la position chromosomique de départ et de fin des nucléotides dupliqués\n",
    "                    if match:\n",
    "                        small_pos = int(match.group(1)) # Petite position chromosomique \n",
    "                        big_pos = int(match.group(2)) # Grande position chromosomique\n",
    "                        # Conversion des positions chromosomiques en position chronologique\n",
    "                        chrono_1 = (get_position_chronologique(gene, small_pos, transcrit)-1)\n",
    "                        chrono_2 = (get_position_chronologique(gene, big_pos, transcrit)-1)\n",
    "                        liste_position_chrono = [chrono_1, chrono_2]\n",
    "                        biggest = max(liste_position_chrono) # Grande position chronologique: fin de la mutation\n",
    "                        smallest = min(liste_position_chrono) # Petite position chronologique: début de la mutation\n",
    "                        sequence_dupliquee = sequence[smallest: biggest+1]\n",
    "                        if sequence_dupliquee == add: # Vérification\n",
    "                            sequence_modifiee = sequence[:biggest+1] + sequence_dupliquee + sequence[biggest+1:] # Formation de la séquence mutées\n",
    "                            header = '>' + f'{gene}_{transcrit}_{mutation_id}' # Formation du header\n",
    "                            fasta = f'../result/duplication.fasta'\n",
    "                            with open (fasta, mode = 'a') as f:\n",
    "                                f.write(header + '\\n') # Écriture du header\n",
    "                                f.write(sequence_modifiee + '\\n') # Écriture de la séquence\n",
    "                        # En théorie, si aucun problème n'est rencontré, la section suivante ne devrait pas s'effectuer\n",
    "                        else: \n",
    "                            if line not in lignes_deja_ajoutees:\n",
    "                                reste = f'../result/mutation_not_working.tsv'\n",
    "                                with open(reste, mode = 'a') as r: # Ouverture du fichier qui va contenir les mutations défectueuses en mode append()\n",
    "                                    r.write(line) # Écriture de la ligne de la mutation\n",
    "                                    lignes_deja_ajoutees.add(line)\n",
    "                else:\n",
    "                    # 1 NUCLÉOTIDE DUPPLIQUÉ\n",
    "                    if not '_' in mutation_explain:\n",
    "                        match = re.search(r'g\\.(\\d+)dup', mutation_explain) # Recherche la position chromosomique du nucléotide qui sera dupliqué\n",
    "                        if match:\n",
    "                            position = int(match.group(1)) # Position chromosomique du nucléotide dupliqué\n",
    "                            chrono_position = int(get_position_chronologique(gene, position, transcrit)-1)\n",
    "                            sequence_dupliquee = sequence[chrono_position]\n",
    "                            if sequence_dupliquee == add: # Vérification\n",
    "                                sequence_modifiee = sequence[:chrono_position+1] + sequence_dupliquee + sequence[chrono_position+1:] # Formation de la séquence mutée\n",
    "                                header = '>' + f'{gene}_{transcrit}_{mutation_id}' # Formation du header\n",
    "                                fasta = f'../result/duplication.fasta' \n",
    "                                with open (fasta, mode = 'a') as f:\n",
    "                                    f.write(header + '\\n') # Écriture du header\n",
    "                                    f.write(sequence_modifiee + '\\n') # Écriture de la séquence\n",
    "                            # En théorie, si aucun problème n'est rencontré, la section suivante ne devrait pas s'effectuer\n",
    "                            else:\n",
    "                                if line not in lignes_deja_ajoutees: \n",
    "                                    reste = f'../result/mutation_not_working.tsv'\n",
    "                                    with open(reste, mode = 'a') as r: # Ouverture du fichier qui va contenir les mutations défectueuses\n",
    "                                        r.write(line) # Écriture de la ligne de la mutation\n",
    "                                        lignes_deja_ajoutees.add(line)\n",
    "\n",
    "            # GÉRER LES DELETION-INSERTION\n",
    "            elif 'delins' in changement:\n",
    "                # Traitement de la séquence deletée\n",
    "                smallest = min(liste) # Petite position chronologique = début de la mutation\n",
    "                biggest = max(liste) # Grande position chronologique = fin de la mutation\n",
    "                remove_sequence = sequence[smallest:biggest+1]\n",
    "\n",
    "                # Traitement de la séquence ajoutée\n",
    "                pattern = r'[ATCG]+'\n",
    "                match = re.search(pattern, changement) # Recherche la séquence ajoutée\n",
    "                sequence_ajoutee = match.group() \n",
    "\n",
    "                if remove_sequence == delete and sequence_ajoutee == add: # Vérification\n",
    "                    sequence_modifiee = sequence[:smallest] + sequence_ajoutee + sequence[biggest+1:] # Formation de la séquence mutée\n",
    "                    header = '>' + f'{gene}_{transcrit}_{mutation_id}' # Formation du header\n",
    "                    fasta = f'../result/del_ins.fasta'\n",
    "                    with open (fasta, 'a') as f: \n",
    "                        f.write(header + '\\n') # Écriture du header\n",
    "                        f.write(sequence_modifiee + '\\n') # Écriture de la séquence\n",
    "                # En théorie, si aucun problème n'est rencontré, la section suivante ne devrait pas s'effectuer\n",
    "                else: \n",
    "                    if line not in lignes_deja_ajoutees:\n",
    "                        reste = f'../result/mutation_not_working.tsv'\n",
    "                        with open(reste, mode = 'a') as r: # Ouverture du fichier qui va contenir les mutations défectueuses en mode append()\n",
    "                            r.write(line) # Écriture de la ligne de la mutation\n",
    "                            lignes_deja_ajoutees.add(line)\n",
    "\n",
    "            # GÉRER LES INVERSIONS\n",
    "            elif changement.endswith('inv'):\n",
    "                smallest = min(liste) # Petite position chronologique = début de la mutation\n",
    "                biggest = max(liste) # Grande position chronologique = fin de la mutation\n",
    "                sequence_inverse = sequence[smallest: biggest+1] \n",
    "                if sequence_inverse == delete: # Vérification\n",
    "                    sequence_modifiee = sequence[:smallest] + sequence_inverse[::-1] + sequence[biggest+1:] # Formation de la séquence mutée\n",
    "                    header = '>' + f'{gene}_{transcrit}_{mutation_id}' # Formation du header\n",
    "                    fasta = f'../result/inversion.fasta' \n",
    "                    with open (fasta, 'a') as f:\n",
    "                        f.write(header + '\\n') # Écriture du header\n",
    "                        f.write(sequence_modifiee + '\\n') # Écriture du header\n",
    "                # En théorie, si aucun problème n'est rencontré, la section suivante ne devrait pas s'effectuer\n",
    "                else:\n",
    "                    if line not in lignes_deja_ajoutees:\n",
    "                        reste = f'../result/mutation_not_working.tsv'\n",
    "                        with open(reste, 'a') as r: # Ouverture du fichier qui va contenir les mutations défectueuses en mode append()\n",
    "                            r.write(line) # Écriture de la ligne de la mutation\n",
    "                            lignes_deja_ajoutees.add(line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "a95de18c",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ÉTAPE 8: AJOUTER TOUS LES TRANSCRITS MUTÉS DANS UN GROS FASTA #####\n",
    "# Cette section permet de combiner tous les sous-fichier fasta qui contiennent les séquences mutés et non-muté classées selon chaque type de mutation, dans\n",
    "# un seul gros fichier fasta.\n",
    "    \n",
    "# Liste de tous les fichiers fasta voulant être combiné\n",
    "fichiers_fasta = ['../result/substitution.fasta', '../result/insertion.fasta', '../result/deletion.fasta', '../result/duplication.fasta', '../result/del_ins.fasta', '../result/inversion.fasta', '../result/transcrit_no_mutation.fasta']\n",
    "# Le nom du fichier qui va contenir tous les contenus\n",
    "fichier_sortie = '../result/all_transcripts_with_mutation.fasta'\n",
    "\n",
    "with open(fichier_sortie, 'w') as fichier_combinaison: # Ouverture du fichier qui va contenir tous les contenus\n",
    "    for fichier in fichiers_fasta:\n",
    "        try:\n",
    "            with open(fichier, 'r') as f: \n",
    "                fichier_combinaison.write(f.read()) # Écrit le contenu du fichier dans le gros fichier\n",
    "        except FileNotFoundError: # Cela se peut, si exemple, le gène ne contient pas de mutation inversion\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "665384f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Correct pour PRDM16 5133==5133\n",
      "Correct pour THRAP3 1463==1463\n",
      "Correct pour CDKN2C 482==482\n",
      "Correct pour TENT5C 490==490\n",
      "Correct pour TPR 2042==2042\n",
      "Correct pour PRRX1 1090==1090\n",
      "Correct pour ARHGEF10L 2698==2698\n"
     ]
    }
   ],
   "source": [
    "# ÉTAPE DE VÉRIFICATION\n",
    "# Le nombre de transcrit obtenu POUR UN GÈNE est égal au nombre de toutes les mutations qui s'effectue pour chaque transcrit + tous les transcrits\n",
    "# à l'état wild-type - le nombre de mutation erroné qui ne s'effectuent pas (doit être compté pour chaque transcrit pour lequel la mutation \n",
    "# s'appliquait)\n",
    "\n",
    "\n",
    "for gene_name in gene_dic2:\n",
    "    count = 0 # Compte le nombre de transcrit pour un gène\n",
    "    count_defectueux = 0 # Compte le nombre de fois qu'une mutation erroné ne s'est pas effectuée\n",
    "    gene = equivalence_nom[gene_name]\n",
    "    \n",
    "    # Compte combien de transcrits ont été produit pour chaque gène\n",
    "    with open('../result/all_transcripts_with_mutation.fasta', 'r') as file:\n",
    "        for line in file:\n",
    "            if gene in line:\n",
    "                count += 1\n",
    "    \n",
    "    # Compte combien de fois une mutation erronée est retrouvé dans un gène\n",
    "    for line in lignes_deja_ajoutees: # Le set contient les lignes erronées\n",
    "        colonne = line.split('\\t')\n",
    "        mutation_id = colonne[8]\n",
    "        if colonne[0] == gene_name:\n",
    "            for transcrit in gene_dic2[gene_name]:\n",
    "                for item in count_les_transcrit_mute_par_gene[gene_name][transcrit]:\n",
    "                    col = item.split('\\t')\n",
    "                    if col[8]== mutation_id:\n",
    "                        count_defectueux += 1\n",
    "                        print(count_defectueux)\n",
    "    \n",
    "    # Compte en théorie combien on devrait avoir de transcrit muté        \n",
    "    nombre_de_mutation = sum(len(liste) for liste in count_les_transcrit_mute_par_gene[gene_name].values())\n",
    "    total = nombre_de_mutation + len(gene_dic2[gene_name])\n",
    "    \n",
    "    if (nombre_de_mutation + len(gene_dic2[gene_name]))== count-count_defectueux:\n",
    "        print(f'Correct pour {gene_name} ' + str(total) + '==' + str(count-count_defectueux))\n",
    "    else:\n",
    "        print(f'Error pour {gene_name} ' + str(total) + '!=' + str(count-count_defectueux))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cosmics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
